<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>profscore &mdash; profscore: Profile Conservation Scoring 1.0.8 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0.8',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="profscore: Profile Conservation Scoring 1.0.8 documentation" href="../index.html" />
    <link rel="up" title="Module code" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">profscore: Profile Conservation Scoring 1.0.8 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for profscore</h1><div class="highlight"><pre>
<span class="c">#-------------------------------------------------------------------------------</span>
<span class="c"># Name:        profscore</span>
<span class="c">#</span>
<span class="c"># Purpose:     Calculates Valdar and profile-based conservation score, with</span>
<span class="c">#              p-values estimated by random sampling from a larger set of</span>
<span class="c">#              sequences.</span>
<span class="c">#</span>
<span class="c"># Author:      Tet Woo Lee &lt;tw.lee@auckland.ac.nz&gt;</span>
<span class="c">#</span>
<span class="c"># Created:     29/03/2011</span>
<span class="c">#</span>
<span class="c"># Copyright:   (c) Tet Woo Lee 2011-2014</span>
<span class="c">#</span>
<span class="c"># Licence:     GNU General Public License version 3</span>
<span class="c">#</span>
<span class="c">#              This program is free software: you can redistribute it and/or modify</span>
<span class="c">#              it under the terms of the GNU General Public License as published by</span>
<span class="c">#              the Free Software Foundation, either version 3 of the License, or</span>
<span class="c">#              (at your option) any later version.</span>
<span class="c">#</span>
<span class="c">#              This program is distributed in the hope that it will be useful,</span>
<span class="c">#              but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c">#              MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c">#              GNU General Public License for more details.</span>
<span class="c">#</span>
<span class="c">#              You should have received a copy of the GNU General Public License</span>
<span class="c">#              along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>
<span class="c">#-------------------------------------------------------------------------------</span>

<span class="c"># Tested with Python 2.7.6 (default, Nov 10 2013, 19:24:24) [MSC v.1500 64 bit (AMD64)]</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">binascii</span> <span class="c"># used for calculating checksum only</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">total_ordering</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">operator</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">time</span>     <span class="c"># used for getting seed for PRNG</span>
<span class="kn">import</span> <span class="nn">timeit</span>
<span class="kn">import</span> <span class="nn">traceback</span>

<span class="c"># Tested with BioPython 1.63 amd64 py2.7</span>
<span class="kn">import</span> <span class="nn">Bio.SeqIO</span>
<span class="kn">import</span> <span class="nn">Bio.SubsMat</span>
<span class="kn">import</span> <span class="nn">Bio.SubsMat.MatrixInfo</span>
<span class="kn">import</span> <span class="nn">Bio.Alphabet</span>

<span class="c"># Tested with numpy 1.9.0 amd64 py2.7</span>
<span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">array</span>
<span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">float64</span>
<span class="kn">import</span> <span class="nn">numpy</span>



<span class="c">##########################################################################################</span>
<span class="c"># GENERIC HELPER FUNCTIONS</span>
<span class="c">##########################################################################################</span>

<div class="viewcode-block" id="createRecordChecksum"><a class="viewcode-back" href="../index.html#profscore.createRecordChecksum">[docs]</a><span class="k">def</span> <span class="nf">createRecordChecksum</span><span class="p">(</span><span class="n">records</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Generated a crc32 checksum of a list of sequence records.</span>
<span class="sd">  Used to uniquely identify a list of sequences.</span>

<span class="sd">  Arguments:</span>

<span class="sd">    records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of records to process</span>

<span class="sd">  Return:</span>

<span class="sd">    out : number</span>
<span class="sd">        32-bit checksum of sequences</span>

<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">checksum</span> <span class="o">=</span> <span class="mi">0</span>
  <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
      <span class="n">crc32</span> <span class="o">=</span> <span class="n">binascii</span><span class="o">.</span><span class="n">crc32</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
      <span class="n">checksum</span><span class="o">+=</span><span class="n">crc32</span>
  <span class="k">return</span> <span class="n">checksum</span> <span class="o">&amp;</span> <span class="mh">0xffffffff</span> <span class="c"># &amp; forces unsigned</span>
</div>
<div class="viewcode-block" id="checkTotalWeight"><a class="viewcode-back" href="../index.html#profscore.checkTotalWeight">[docs]</a><span class="k">def</span> <span class="nf">checkTotalWeight</span><span class="p">(</span><span class="n">records</span><span class="p">,</span><span class="n">weights</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Return ``True`` if sum of weights adds up to ``1.0``.</span>

<span class="sd">  Arguments:</span>

<span class="sd">    records : list of Bio.SeqRecord objects</span>
<span class="sd">      list of aligned sequences to test weights for</span>
<span class="sd">    weights : NamedDict of sequence weights</span>
<span class="sd">      sequences weights to check</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">totalWeight</span> <span class="o">=</span> <span class="mf">0.0</span>
  <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
      <span class="n">weight</span> <span class="o">=</span> <span class="n">weights</span><span class="p">[</span><span class="n">record</span><span class="p">]</span>
      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span> <span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s">&#39;has weight&#39;</span><span class="p">,</span><span class="n">weight</span>
      <span class="n">totalWeight</span><span class="o">+=</span><span class="n">weight</span>

  <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">3</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Check sequence weights: Total weight is&#39;</span><span class="p">,</span><span class="n">totalWeight</span>
  <span class="k">return</span> <span class="nb">abs</span><span class="p">(</span><span class="n">totalWeight</span><span class="o">-</span><span class="mf">1.0</span><span class="p">)</span><span class="o">&lt;</span><span class="mf">1e-9</span>
</div>
<div class="viewcode-block" id="NamedDict"><a class="viewcode-back" href="../index.html#profscore.NamedDict">[docs]</a><span class="k">class</span> <span class="nc">NamedDict</span><span class="p">(</span><span class="nb">dict</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Simple subclass of ``dict`` that contains a ``name`` attributed.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s">&#39;name&#39;</span><span class="p">)</span> <span class="c"># use slots to conserve memory</span>
  <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">name</span><span class="p">):</span>
    <span class="nb">dict</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span> <span class="c">#: *string*, name of this dict</span>
</div>
<div class="viewcode-block" id="printRecordNames"><a class="viewcode-back" href="../index.html#profscore.printRecordNames">[docs]</a><span class="k">def</span> <span class="nf">printRecordNames</span><span class="p">(</span><span class="n">records</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Simple function to print record ids.</span>

<span class="sd">  Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of records to print record ids</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="k">print</span> <span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="isProfileScore"><a class="viewcode-back" href="../index.html#profscore.isProfileScore">[docs]</a><span class="k">def</span> <span class="nf">isProfileScore</span><span class="p">(</span><span class="n">scoreType</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Return ``True`` if ``scoretype`` is a profile-based score.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="k">return</span> <span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE</span> <span class="ow">or</span> <span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE_UNCORR</span>

</div>
<div class="viewcode-block" id="subsetByID"><a class="viewcode-back" href="../index.html#profscore.subsetByID">[docs]</a><span class="k">def</span> <span class="nf">subsetByID</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">nameREstr</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return a subset of sequences by regular expression search.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">      nameREstr : string</span>
<span class="sd">        regular expression to use to subset sequences, sequences must contain this regular expression</span>


<span class="sd">    Return:</span>

<span class="sd">      out : list of Bio.SeqRecord objects</span>
<span class="sd">        list of sequences containing the regex</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">print</span> <span class="s">&quot;Creating a subset of records by searching for regex </span><span class="si">%s</span><span class="s"> in id...&quot;</span><span class="o">%</span><span class="n">nameREstr</span>
    <span class="n">nameRE</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">nameREstr</span><span class="p">)</span>
    <span class="n">sample</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">nameRE</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Sequence </span><span class="si">%s</span><span class="s"> </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="s">&quot;contains regex&quot;</span> <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="k">else</span> <span class="s">&quot;does not contain regex&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">sample</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sample</span>
</div>
<div class="viewcode-block" id="enum"><a class="viewcode-back" href="../index.html#profscore.enum">[docs]</a><span class="k">def</span> <span class="nf">enum</span><span class="p">(</span><span class="o">*</span><span class="n">sequential</span><span class="p">,</span> <span class="o">**</span><span class="n">named</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Simple method to create enum-like objects.</span>

<span class="sd">  Usage: ``enum(&#39;APPLE&#39;,&#39;ORANGE&#39;,&#39;PEAR&#39;)``</span>

<span class="sd">  Source: Alec Thomas, stack overflow http://stackoverflow.com/questions/36932/how-can-i-represent-an-enum-in-python</span>

<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">enums</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">sequential</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequential</span><span class="p">))),</span> <span class="o">**</span><span class="n">named</span><span class="p">)</span>
  <span class="n">reverse</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">value</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">enums</span><span class="o">.</span><span class="n">iteritems</span><span class="p">())</span>
  <span class="n">enums</span><span class="p">[</span><span class="s">&#39;reverse_mapping&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reverse</span>
  <span class="k">return</span> <span class="nb">type</span><span class="p">(</span><span class="s">&#39;Enum&#39;</span><span class="p">,</span> <span class="p">(),</span> <span class="n">enums</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="Dummy"><a class="viewcode-back" href="../index.html#profscore.Dummy">[docs]</a><span class="k">class</span> <span class="nc">Dummy</span><span class="p">:</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Dummy object class for storing data.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="k">pass</span>

<span class="c">##########################################################################################</span>
<span class="c"># ENUMS</span>
<span class="c">##########################################################################################</span>
</div>
<span class="n">ScoreType</span> <span class="o">=</span> <span class="n">enum</span><span class="p">(</span><span class="s">&#39;VALDAR&#39;</span><span class="p">,</span><span class="s">&#39;PROFILE&#39;</span><span class="p">,</span><span class="s">&#39;PROFILE_UNCORR&#39;</span><span class="p">)</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Simple enum-like class for specifying score type.</span>

<span class="sd">Possible values:</span>

<span class="sd">  ``VALDAR``</span>
<span class="sd">    Valdar score</span>
<span class="sd">  ``PROFILE``</span>
<span class="sd">    Profile-based score, corrected for self-comparisons</span>
<span class="sd">  ``PROFILE_UNCORR``</span>
<span class="sd">    Profile-based score, not corrected for self-comparisons</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="n">FilePosition</span> <span class="o">=</span> <span class="n">enum</span><span class="p">(</span><span class="s">&#39;START&#39;</span><span class="p">,</span><span class="s">&#39;HEADER&#39;</span><span class="p">,</span><span class="s">&#39;DATA&#39;</span><span class="p">,</span><span class="s">&#39;END&#39;</span><span class="p">)</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Simple enum-like class for specifying file position when reading randomization file.</span>

<span class="sd">Possible values:</span>

<span class="sd">  ``START``</span>
<span class="sd">    Initial state, before reading magic start word</span>
<span class="sd">  ``HEADER``</span>
<span class="sd">    Reading header with parameterrs</span>
<span class="sd">  ``DATA``</span>
<span class="sd">    Reading data site by site</span>
<span class="sd">  ``END``</span>
<span class="sd">    After reading magic end word</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="n">OutputPositions</span> <span class="o">=</span> <span class="n">enum</span><span class="p">(</span><span class="s">&#39;ALL_SIMPLENUM&#39;</span><span class="p">,</span><span class="s">&#39;ALL&#39;</span><span class="p">,</span><span class="s">&#39;PRESENT&#39;</span><span class="p">,</span><span class="s">&#39;REFERENCE&#39;</span><span class="p">)</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Simple enum-like class for specifying output positions when reading reports.</span>

<span class="sd">Possible values:</span>

<span class="sd">  ``ALL``</span>
<span class="sd">    output all positions in alignment</span>
<span class="sd">  ``PRESENT``</span>
<span class="sd">    output only present positions in subset</span>
<span class="sd">  ``REFERENCE``</span>
<span class="sd">    output only positions present in reference sequence</span>

<span class="sd">&quot;&quot;&quot;</span>


<span class="c">##########################################################################################</span>
<span class="c"># SUBSTITUTION MATRIX FUNCTIONS</span>
<span class="c">##########################################################################################</span>

<div class="viewcode-block" id="readSubsMat"><a class="viewcode-back" href="../index.html#profscore.readSubsMat">[docs]</a><span class="k">def</span> <span class="nf">readSubsMat</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read a substitution matrix from file in aaindex2 format.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      filename : string</span>
<span class="sd">          path to the file to read, file should be in aaindex2 format</span>

<span class="sd">    Return:</span>

<span class="sd">      out : Bio.SubsMat.SeqMat object</span>
<span class="sd">          substitution matrix read from file</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">infile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&#39;Reading substitution matrix:&#39;</span><span class="p">,</span><span class="n">filename</span><span class="p">,</span><span class="s">&#39;...&#39;</span>
    <span class="n">row</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="c"># -1 = reading header, &gt;=0 reading matrix</span>
    <span class="n">desc</span>  <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="c"># description from &#39;D&#39; line</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">row</span><span class="o">&lt;</span><span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;D&#39;</span><span class="p">:</span> <span class="c"># D = description</span>
                <span class="n">desc</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                <span class="k">print</span> <span class="s">&#39;Description:&#39;</span><span class="p">,</span> <span class="n">desc</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;M&#39;</span><span class="p">:</span> <span class="c"># M = matrix info (i.e. letters)</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">,</span><span class="s">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">,</span><span class="s">&#39; &#39;</span><span class="p">)</span>
                <span class="n">Mtokens</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">rowstoken</span> <span class="o">=</span> <span class="n">Mtokens</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s">&#39;rows&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span>
                <span class="n">colstoken</span> <span class="o">=</span> <span class="n">Mtokens</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s">&#39;cols&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span>
                <span class="n">rows</span> <span class="o">=</span> <span class="n">Mtokens</span><span class="p">[</span><span class="n">rowstoken</span><span class="p">]</span>
                <span class="n">cols</span> <span class="o">=</span> <span class="n">Mtokens</span><span class="p">[</span><span class="n">colstoken</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Rows:&#39;</span><span class="p">,</span> <span class="n">rows</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Cols:&#39;</span><span class="p">,</span> <span class="n">rows</span>
                <span class="k">if</span> <span class="n">rows</span><span class="o">!=</span><span class="n">cols</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;rows and cols must be identical&#39;</span><span class="p">)</span>
                <span class="n">seqmat</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SubsMat</span><span class="o">.</span><span class="n">SeqMat</span><span class="p">(</span><span class="n">mat_name</span> <span class="o">=</span> <span class="n">desc</span><span class="p">)</span>
                <span class="n">seqmat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span> <span class="o">=</span> <span class="n">rows</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Reading matrix...&#39;</span>
                <span class="n">row</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">entries</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">entries</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;//&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Reached end of matrix.&#39;</span>
                <span class="n">seqmat</span><span class="o">.</span><span class="n">_correct_matrix</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                  <span class="n">seqmat</span><span class="o">.</span><span class="n">print_mat</span><span class="p">(</span> <span class="n">format</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%5.1f</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">bottomformat</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%5s</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">seqmat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">)</span>
                <span class="k">break</span>
            <span class="n">row_letter</span> <span class="o">=</span> <span class="n">rows</span><span class="p">[</span><span class="n">row</span><span class="p">]</span>
            <span class="n">col</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">entries</span><span class="p">:</span>
                <span class="n">col_letter</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="n">col</span><span class="p">]</span>
                <span class="n">seqmat</span><span class="p">[</span><span class="n">col_letter</span><span class="p">,</span><span class="n">row_letter</span><span class="p">]</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="n">entry</span><span class="p">)</span>
                <span class="n">col</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">row</span><span class="o">+=</span><span class="mi">1</span>

    <span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">seqmat</span>
</div>
<div class="viewcode-block" id="linearTransform"><a class="viewcode-back" href="../index.html#profscore.linearTransform">[docs]</a><span class="k">def</span> <span class="nf">linearTransform</span><span class="p">(</span><span class="n">subsMat</span><span class="p">,</span> <span class="n">diagonal</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">gap</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span> <span class="n">gapgap</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Linear transform a substitution matrix, so that min is 0.0 and max is 1.0.</span>
<span class="sd">    Note that the matrix is transformed in place.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      subsMat : Bio.SubsMat.SeqMat object</span>
<span class="sd">        the substitution matrix to transform</span>
<span class="sd">      diagonal : number, optional</span>
<span class="sd">        value given to diagonal elements in matrix, if ``None`` (default) diagonal to set to &quot;rounded average of diagonal elements in the unmodified matrix&quot; (Valdar 2002),</span>
<span class="sd">        applied before transformation</span>
<span class="sd">      gap : number, optional</span>
<span class="sd">        score given to aa -&gt; gap substitutions, defaults to ``0.0``, applied after transformation</span>
<span class="sd">      gapgap : number, optional</span>
<span class="sd">        score for gap -&gt; gap substitutions, if ``None``, same as gap score, applied after transformation</span>

<span class="sd">    Return:</span>

<span class="sd">      out : Bio.SubsMat.SeqMat object</span>
<span class="sd">        Transformed substitution matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">print</span> <span class="s">&#39;Linear transforming matrix:&#39;</span><span class="p">,</span><span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span>

    <span class="k">if</span> <span class="n">gapgap</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span> <span class="n">gapgap</span> <span class="o">=</span> <span class="n">gap</span>

    <span class="c"># transform diagonal to &quot;rounded average of diagonal elements in the unmodified matrix&quot; (Valdar 2002)</span>
    <span class="n">aaletters</span> <span class="o">=</span> <span class="s">&#39;ARNDCQEGHILKMFPSTWYV&#39;</span> <span class="c"># use only aa letters</span>
    <span class="k">if</span> <span class="n">diagonal</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">diagonalSum</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">aaletters</span><span class="p">:</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">a</span><span class="p">,</span><span class="n">a</span><span class="p">]</span>
            <span class="n">diagonalSum</span><span class="o">+=</span><span class="n">m</span>
            <span class="n">n</span><span class="o">+=</span><span class="mi">1</span>
        <span class="n">diagonalScore</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">diagonalSum</span><span class="o">/</span><span class="n">n</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span><span class="o">+=</span><span class="s">&#39;, average diagonal, linear transform, gap </span><span class="si">%.3f</span><span class="s">, gap-gap </span><span class="si">%.3f</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">gap</span><span class="p">,</span><span class="n">gapgap</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">diagonalScore</span> <span class="o">=</span> <span class="n">diagonal</span>
        <span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span><span class="o">+=</span><span class="s">&#39;, diagonal </span><span class="si">%.3f</span><span class="s">, linear transform, gap </span><span class="si">%.3f</span><span class="s">, gap-gap </span><span class="si">%.3f</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">diagonal</span><span class="p">,</span><span class="n">gap</span><span class="p">,</span><span class="n">gapgap</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Diagonal score set to:&#39;</span><span class="p">,</span><span class="n">diagonalScore</span>
    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">:</span>
        <span class="n">subsMat</span><span class="p">[</span><span class="n">a</span><span class="p">,</span><span class="n">a</span><span class="p">]</span> <span class="o">=</span> <span class="n">diagonalScore</span>
    <span class="nb">min</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">aaletters</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">aaletters</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="nb">max</span> <span class="o">=</span> <span class="nb">min</span>
    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">aaletters</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">aaletters</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">a</span><span class="o">&lt;</span><span class="n">b</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">m</span><span class="o">&gt;</span><span class="nb">max</span><span class="p">:</span> <span class="nb">max</span> <span class="o">=</span> <span class="n">m</span>
                <span class="k">if</span> <span class="n">m</span><span class="o">&lt;</span><span class="nb">min</span><span class="p">:</span> <span class="nb">min</span> <span class="o">=</span> <span class="n">m</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Matrix min:&#39;</span><span class="p">,</span><span class="nb">min</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Matrix max:&#39;</span><span class="p">,</span><span class="nb">max</span>
    <span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="o">+=</span><span class="s">&#39;-&#39;</span>
    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">a</span><span class="o">&lt;=</span><span class="n">b</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">a</span><span class="o">==</span><span class="s">&#39;-&#39;</span> <span class="ow">and</span> <span class="n">b</span><span class="o">==</span><span class="s">&#39;-&#39;</span><span class="p">:</span>
                    <span class="n">Mut</span> <span class="o">=</span> <span class="n">gapgap</span>
                <span class="k">elif</span> <span class="n">a</span><span class="o">==</span><span class="s">&#39;-&#39;</span><span class="p">:</span>
                    <span class="n">Mut</span> <span class="o">=</span> <span class="n">gap</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">m</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">]</span>
                    <span class="n">Mut</span> <span class="o">=</span> <span class="p">(</span><span class="n">m</span> <span class="o">-</span> <span class="nb">min</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">max</span> <span class="o">-</span> <span class="nb">min</span><span class="p">)</span>
                <span class="n">subsMat</span><span class="p">[</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">]</span>  <span class="o">=</span> <span class="n">Mut</span>
                <span class="n">subsMat</span><span class="p">[</span><span class="n">b</span><span class="p">,</span><span class="n">a</span><span class="p">]</span>  <span class="o">=</span> <span class="n">Mut</span>
    <span class="k">print</span> <span class="s">&quot;New matrix:&quot;</span><span class="p">,</span><span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="n">subsMat</span><span class="o">.</span><span class="n">print_mat</span><span class="p">(</span> <span class="n">format</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%6.3f</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">bottomformat</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%6s</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="VectorSubsMat"><a class="viewcode-back" href="../index.html#profscore.VectorSubsMat">[docs]</a><span class="k">class</span> <span class="nc">VectorSubsMat</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a substitution matrix in vector form from a (standard) substitution matrix.</span>
<span class="sd">    If ``vectorSubsMat`` is an instance of this class, then each lookup of ``vectorSubsMat[aa1]``</span>
<span class="sd">    (where ``aa1`` is amino acid in upper case one letter code)</span>
<span class="sd">    provides a *k*-dimensional vector giving scores for all possible substitutions.</span>

<span class="sd">    If ``vector = vectorSubsMat[aa1]``, a particular substitution score get be found by</span>
<span class="sd">    ``vector[index]`` where ``index`` is  the index of a (second) aa.</span>
<span class="sd">    The index of an aa given by ``vectorSubsMat.alphabetLUT[ord(aa2)]``.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      subsMat : Bio.SubsMat.SeqMat object</span>
<span class="sd">        the substitution matrix to convert</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subsMat</span><span class="p">):</span>
        <span class="k">print</span> <span class="s">&#39;Creating vector substitution matrix from:&#39;</span><span class="p">,</span><span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mat_name</span> <span class="o">=</span> <span class="s">&#39;vectors of &#39;</span><span class="o">+</span><span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span> <span class="c">#: *string*, name of this substitution matrix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Alphabet</span><span class="o">.</span><span class="n">Alphabet</span><span class="p">()</span> <span class="c">#: *Bio.Alphabet.Alphabet object*, alphabet of this substitution matrix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span> <span class="o">=</span> <span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span> <span class="c">#: *string*, letters of this alphabet</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span> <span class="o">=</span> <span class="n">subsMat</span> <span class="c">#: *Bio.SubsMat.SeqMat object*, source substitution matrix</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">scoreLUT</span> <span class="o">=</span> <span class="p">{}</span> <span class="c">#: *dict*, look-up table mapping amino acid letters to vectors, access by indexing ``VectorSubsMat`` object, i.e. ``vectorSubsMat[aa] = scoreVector``</span>

        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">:</span>
            <span class="n">scores</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="n">float64</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">bi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">)):</span>
                <span class="n">b</span> <span class="o">=</span> <span class="n">subsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">[</span><span class="n">bi</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">a</span><span class="o">&lt;=</span><span class="n">b</span><span class="p">:</span> <span class="c"># use lower triangle only</span>
                    <span class="n">m</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">m</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">b</span><span class="p">,</span><span class="n">a</span><span class="p">]</span>
                <span class="n">scores</span><span class="p">[</span><span class="n">bi</span><span class="p">]</span> <span class="o">=</span>  <span class="n">m</span>
            <span class="c">#VectorSubsMat[a] = scores</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">scoreLUT</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="o">=</span> <span class="n">scores</span>


        <span class="c"># generate lookup table to get alphabet_index from ordinal</span>
        <span class="n">maxOrdinal</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">:</span>
            <span class="n">ordinal</span> <span class="o">=</span> <span class="nb">ord</span><span class="p">(</span><span class="n">letter</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ordinal</span><span class="o">&gt;</span><span class="n">maxOrdinal</span><span class="p">:</span> <span class="n">maxOrdinal</span> <span class="o">=</span> <span class="n">ordinal</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">alphabetLUT</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">maxOrdinal</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        *list*, look-up table of amino acid letters to substitution vector indexes.</span>

<span class="sd">        Usage:</span>
<span class="sd">          ``alphabetLUT[ord(aa)]`` where ``aa`` is the amino acid</span>
<span class="sd">          in upper-case one letter code e.g. ``A``, ``Y``, ``M`` etc.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">letterIndex</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">)):</span>
            <span class="n">letter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">[</span><span class="n">letterIndex</span><span class="p">]</span>
            <span class="n">ordinal</span> <span class="o">=</span> <span class="nb">ord</span><span class="p">(</span><span class="n">letter</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">alphabetLUT</span><span class="p">[</span><span class="n">ordinal</span><span class="p">]</span> <span class="o">=</span> <span class="n">letterIndex</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">dimensions</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">)</span> <span class="c">#: *number*, number of dimensions in vector (i.e. aa characters)</span>


    <span class="k">def</span> <span class="nf">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoreLUT</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>



<span class="c">##########################################################################################</span>
<span class="c"># CALCULATION OF SEQUENCE DISTANCES AND WEIGHTS</span>
<span class="c">##########################################################################################</span>
</div>
<div class="viewcode-block" id="evolDistance"><a class="viewcode-back" href="../index.html#profscore.evolDistance">[docs]</a><span class="k">def</span> <span class="nf">evolDistance</span><span class="p">(</span><span class="n">recordI</span><span class="p">,</span> <span class="n">recordJ</span><span class="p">,</span> <span class="n">subsMat</span><span class="p">):</span>
    <span class="sd">r&quot;&quot;&quot;</span>
<span class="sd">    Calculate the evolutionary distance between two aligned sequence records given a substitution matrix.</span>
<span class="sd">    As per Valdar, 2001 PMID: 11093265</span>

<span class="sd">    .. math::</span>
<span class="sd">            D(s_i,s_j) = 1 - \frac{\displaystyle\sum_{x \in aligned_{ij}}M\Big (s_{i}(x),s_{j}(x)\Big )}{\displaystyle n(aligned_{ij})}</span>

<span class="sd">    Arguments:</span>

<span class="sd">      recordI : Bio.SeqRecord object</span>
<span class="sd">        first sequence</span>
<span class="sd">      recordJ : Bio.SeqRecord object</span>
<span class="sd">        second sequence</span>
<span class="sd">      subsMat : Bio.SubsMat.SeqMat object</span>
<span class="sd">        substitution matrix, transformed</span>

<span class="sd">    Returns:</span>

<span class="sd">      out : number</span>
<span class="sd">        evolutionary distance</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seqI</span> <span class="o">=</span> <span class="n">recordI</span><span class="o">.</span><span class="n">seq</span>
    <span class="n">seqJ</span> <span class="o">=</span> <span class="n">recordJ</span><span class="o">.</span><span class="n">seq</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqI</span><span class="p">)</span><span class="o">!=</span><span class="nb">len</span><span class="p">(</span><span class="n">seqJ</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Sequences must be same length </span><span class="si">%s</span><span class="s"> </span><span class="si">%i</span><span class="s"> vs </span><span class="si">%s</span><span class="s"> </span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">recordI</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">seqI</span><span class="p">),</span><span class="n">recordJ</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">seqJ</span><span class="p">)))</span>
    <span class="nb">sum</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seqI</span><span class="p">)):</span>
        <span class="n">sj</span> <span class="o">=</span> <span class="n">seqI</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">sk</span> <span class="o">=</span> <span class="n">seqJ</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">sj</span><span class="o">==</span><span class="s">&#39;-&#39;</span> <span class="ow">and</span> <span class="n">sk</span><span class="o">==</span><span class="s">&#39;-&#39;</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">Mut</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">sj</span><span class="p">,</span><span class="n">sk</span><span class="p">]</span>
        <span class="c">#print sj,sk,Mut</span>
        <span class="nb">sum</span><span class="o">+=</span><span class="n">Mut</span>
        <span class="n">n</span><span class="o">+=</span><span class="mi">1</span>
    <span class="n">result</span> <span class="o">=</span> <span class="mf">1.0</span><span class="o">-</span><span class="nb">sum</span><span class="o">/</span><span class="n">n</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">3</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Calculated evolutionary distance between </span><span class="si">%s</span><span class="s"> and </span><span class="si">%s</span><span class="s">: </span><span class="si">%f</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">recordI</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="n">recordJ</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="n">result</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">result</span>
</div>
<div class="viewcode-block" id="createUniformDistanceMatrix"><a class="viewcode-back" href="../index.html#profscore.createUniformDistanceMatrix">[docs]</a><span class="k">def</span> <span class="nf">createUniformDistanceMatrix</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">uniformDistance</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a simple distance matrix with a uniform value for pairs of aligned sequences.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences (only size of list is used by this function)</span>
<span class="sd">      uniformDistance : number, optional</span>
<span class="sd">        the value to use as the uniform distance between each pair of sequences, defaults to ``1.0``</span>

<span class="sd">    Returns:</span>

<span class="sd">      out : NamedDict</span>
<span class="sd">        A dict of giving uniform distance for each pair of records *ij* i.e ``out[recordI, recordJ] = uniformDistance``.</span>
<span class="sd">        Value of ``out[recordI, recordI]`` is undefined.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Creating uniform evolutionary distance matrix...&quot;</span>
    <span class="n">uniformDistance</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="n">uniformDistance</span><span class="p">)</span>
    <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">NamedDict</span><span class="p">(</span><span class="s">&quot;uniform distance of </span><span class="si">%f</span><span class="s">&quot;</span><span class="o">%</span><span class="n">uniformDistance</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">recordJ</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">recordJ</span><span class="o">&gt;</span><span class="n">recordI</span><span class="p">:</span>
                <span class="n">distanceMatrix</span><span class="p">[</span><span class="n">recordI</span><span class="p">,</span> <span class="n">recordJ</span><span class="p">]</span> <span class="o">=</span> <span class="n">uniformDistance</span>
                <span class="n">distanceMatrix</span><span class="p">[</span><span class="n">recordJ</span><span class="p">,</span> <span class="n">recordI</span><span class="p">]</span> <span class="o">=</span> <span class="n">uniformDistance</span>
                <span class="n">n</span><span class="o">+=</span><span class="mi">1</span>
    <span class="k">print</span> <span class="s">&#39;Created uniform distance matrix for </span><span class="si">%i</span><span class="s"> pairwise comparisons&#39;</span><span class="o">%</span><span class="n">n</span>
    <span class="k">return</span> <span class="n">distanceMatrix</span>
</div>
<div class="viewcode-block" id="calculateDistanceMatrix"><a class="viewcode-back" href="../index.html#profscore.calculateDistanceMatrix">[docs]</a><span class="k">def</span> <span class="nf">calculateDistanceMatrix</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">subsMat</span><span class="p">):</span>
    <span class="sd">r&quot;&quot;&quot;</span>
<span class="sd">    Calculate a matrix of evolutionary distances between pairs of aligned sequences given a substitution matrix.</span>
<span class="sd">    Distance calculations as per Valdar, 2001 PMID: 11093265</span>

<span class="sd">    .. math::</span>
<span class="sd">            D(s_i,s_j) = 1 - \frac{\displaystyle\sum_{x \in aligned_{ij}}M\Big (s_{i}(x),s_{j}(x)\Big )}{\displaystyle n(aligned_{ij})}</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">      subsMat : Bio.SubsMat.SeqMat object</span>
<span class="sd">        substitution matrix, transformed</span>

<span class="sd">    Returns:</span>

<span class="sd">      out : NamedDict</span>
<span class="sd">        A dict giving evolutionary distance for each pair of records ij i.e ``out[recordI, recordJ] = distance(recordI, recordJ)``.</span>
<span class="sd">        Matrix is symmetrical and ``out[recordI, recordI]`` is undefined.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Calculated evolutionary distance matrix...&quot;</span>
    <span class="n">startTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">NamedDict</span><span class="p">(</span><span class="s">&quot;evolutionary distance&quot;</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">recordJ</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">recordJ</span><span class="o">&gt;</span><span class="n">recordI</span><span class="p">:</span>
                <span class="n">dist</span> <span class="o">=</span> <span class="n">evolDistance</span><span class="p">(</span><span class="n">recordI</span><span class="p">,</span><span class="n">recordJ</span><span class="p">,</span><span class="n">subsMat</span><span class="p">)</span>
                <span class="n">distanceMatrix</span><span class="p">[</span><span class="n">recordI</span><span class="p">,</span> <span class="n">recordJ</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>
                <span class="n">distanceMatrix</span><span class="p">[</span><span class="n">recordJ</span><span class="p">,</span> <span class="n">recordI</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>
                <span class="n">n</span><span class="o">+=</span><span class="mi">1</span>
    <span class="k">print</span> <span class="s">&#39;Calculated distance matrix for </span><span class="si">%i</span><span class="s"> pairwise comparisons&#39;</span><span class="o">%</span><span class="n">n</span>
    <span class="n">endTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Calculation time was </span><span class="si">%.3f</span><span class="s"> secs.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">endTime</span><span class="o">-</span><span class="n">startTime</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">distanceMatrix</span>
</div>
<div class="viewcode-block" id="calculateWeights"><a class="viewcode-back" href="../index.html#profscore.calculateWeights">[docs]</a><span class="k">def</span> <span class="nf">calculateWeights</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">distanceMatrix</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>

    <span class="sd">r&quot;&quot;&quot;</span>
<span class="sd">    Calculate weight of each sequence *|s_i|* as mean evolutionary distance between *|s_i|* and all other sequences in the alignment.</span>
<span class="sd">    As per Valdar, 2001 PMID: 11093265</span>

<span class="sd">    .. math::</span>
<span class="sd">          w_i = \frac{\displaystyle\sum_{j\neq i}^{N}D(s_i,s_j)}{\displaystyle N-1}</span>
<span class="sd">    .. |s_i| replace:: s\ :sub:`i`</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">      distanceMatrix : NamedDict of evolutionary distances</span>
<span class="sd">        pre-calculated distance matrix giving pairwise distances between all records to process</span>
<span class="sd">      silent : boolean, optional</span>
<span class="sd">        if ``True`` do not produce any output (used when called by randomization functions), defaults to ``False``</span>

<span class="sd">    Returns:</span>

<span class="sd">      out : NamedDict</span>
<span class="sd">        A dict of giving weight for each sequence record, i.e. ``out[recordI] = weightI``</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">silent</span> <span class="ow">and</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Calculating sequence weights from evolutionary distance matrix...&quot;</span>
    <span class="n">startTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="n">weights</span> <span class="o">=</span> <span class="n">NamedDict</span><span class="p">(</span><span class="s">&quot;weighted by </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">distanceMatrix</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="n">N</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">totalWeight</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
        <span class="n">sumDistance</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">recordJ</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">recordI</span><span class="o">==</span><span class="n">recordJ</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">sumDistance</span><span class="o">+=</span><span class="n">distanceMatrix</span><span class="p">[</span><span class="n">recordI</span><span class="p">,</span> <span class="n">recordJ</span><span class="p">]</span>
        <span class="n">weight</span> <span class="o">=</span> <span class="n">sumDistance</span><span class="o">/</span><span class="n">N</span>
        <span class="n">weights</span><span class="p">[</span><span class="n">recordI</span><span class="p">]</span> <span class="o">=</span> <span class="n">weight</span>
        <span class="n">totalWeight</span> <span class="o">+=</span> <span class="n">weight</span>
    <span class="c"># normalise to total 1</span>
    <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
        <span class="n">weight</span> <span class="o">=</span> <span class="n">weights</span><span class="p">[</span><span class="n">recordI</span><span class="p">]</span>
        <span class="n">weight</span> <span class="o">=</span> <span class="n">weight</span><span class="o">/</span><span class="n">totalWeight</span>
        <span class="n">weights</span><span class="p">[</span><span class="n">recordI</span><span class="p">]</span> <span class="o">=</span> <span class="n">weight</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">silent</span> <span class="ow">and</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Weight of sequence </span><span class="si">%s</span><span class="s">: </span><span class="si">%f</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">recordI</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="n">weight</span><span class="p">)</span>
    <span class="n">endTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Calculation time was </span><span class="si">%.3f</span><span class="s"> secs.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">endTime</span><span class="o">-</span><span class="n">startTime</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">weights</span>



<span class="c">##########################################################################################</span>
<span class="c"># CONSERVATION SCORING</span>
<span class="c">##########################################################################################</span>
</div>
<div class="viewcode-block" id="ConservationScores"><a class="viewcode-back" href="../index.html#profscore.ConservationScores">[docs]</a><span class="k">class</span> <span class="nc">ConservationScores</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A data storage class to store calculated sequence conservations scores.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">      scoreName : string</span>
<span class="sd">        name of this scoring system, added to ``self.metadata``</span>
<span class="sd">      scoreAgainst : string</span>
<span class="sd">        details of scoring (e.g. scoring against what matrix/profile), added to ``self.metadata``</span>
<span class="sd">      weightBy: string</span>
<span class="sd">        details of weighting, added to ``self.metadata``</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">scoreName</span><span class="p">,</span> <span class="n">weightBy</span><span class="p">,</span> <span class="n">scoreAgainst</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Empty sequence records list provided!&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">records</span> <span class="o">=</span> <span class="n">records</span> <span class="c">#: *list of Bio.SeqRecord objects*, list of records provided to this score object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="c">#: *number*, alignment length, i.e. total number of sites in alignment</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span> <span class="c">#: *OrderedDict*, metadata for this profile</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">scores</span> <span class="o">=</span> <span class="bp">None</span> <span class="c">#: *list of numbers*, list of scores indexed by 0-based site number in alignment, ``None`` until scores calculated</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span> <span class="c">#: *OrderedDict*, metadata for this profile</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;score name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scoreName</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;score sequences&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%i</span><span class="s"> sequences (</span><span class="si">%08X</span><span class="s"> checksum)&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">),</span><span class="n">createRecordChecksum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;score sequence ids&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;score number of sites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%i</span><span class="s"> sites&quot;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;score weighting&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">weightBy</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;score against&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scoreAgainst</span>

<div class="viewcode-block" id="ConservationScores.allocateScores"><a class="viewcode-back" href="../index.html#profscore.ConservationScores.allocateScores">[docs]</a>    <span class="k">def</span> <span class="nf">allocateScores</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Allocate an array of length ``self.M`` for storing scores, with all values initially set to ``nan``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scores</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">float64</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scores</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span>




<span class="c">##########################################################################################</span>
<span class="c"># VALDAR CONSERVATION SCORING</span>
<span class="c">##########################################################################################</span>
</div></div>
<div class="viewcode-block" id="ValdarScore"><a class="viewcode-back" href="../index.html#profscore.ValdarScore">[docs]</a><span class="k">class</span> <span class="nc">ValdarScore</span><span class="p">(</span><span class="n">ConservationScores</span><span class="p">):</span>
  <span class="sd">r&quot;&quot;&quot;</span>
<span class="sd">  Calculate the Valdar conservation score for all sites in an alignment, given a list of sequence</span>
<span class="sd">  records, a substitution matrix and a distance matrix or list of sequence weights.</span>
<span class="sd">  As per Valdar, 2001 PMID: 11093265</span>

<span class="sd">  Scores are automatically calculated when the class in constructed. Once constructed,</span>
<span class="sd">  this class will contain list of Valdar scores for each site in the alignment in ``self.scores``.</span>

<span class="sd">  Inherits from ``ConservationScores``.</span>

<span class="sd">  Valdar score:</span>

<span class="sd">  .. math::</span>
<span class="sd">     V_x = \frac{\displaystyle \sum_{i}^{N}\sum_{j&gt;i}^{N}w_i w_j M\Big (s_{ix},s_{jx}\Big ) } {\displaystyle \sum_{i}^{N}\sum_{j&gt;i}^{N}w_i w_j }</span>

<span class="sd">  Arguments:</span>

<span class="sd">    records : list of Bio.SeqRecord objects</span>
<span class="sd">      list of aligned sequences to process</span>
<span class="sd">    subsMat : Bio.SubsMat.SeqMat object</span>
<span class="sd">      substitution matrix, transformed</span>
<span class="sd">    distanceMatrix : NamedDict of evolutionary distances, optional, may be ``None``</span>
<span class="sd">      pre-calculated distance matrix giving pairwise distances between all records (if not provided, calculated automatically; not used if weights provided)</span>
<span class="sd">    weights : NamedDict of sequence weights, optional, may be ``None``</span>
<span class="sd">      pre-calculated weights of all sequences (if not provided, calculated automatically from evolutionary distances)</span>

<span class="sd">  &quot;&quot;&quot;</span>
  <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">subsMat</span><span class="p">,</span> <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">weights</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Calculating Valdar conservation scores...&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">distanceMatrix</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">weights</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Either distanceMatrix or weights must be provided&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span> <span class="o">=</span> <span class="n">subsMat</span> <span class="c">#: *Bio.SubsMat.SeqMat object*, substitution matrix used to calculated scores</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">weights</span> <span class="o">=</span> <span class="n">weights</span> <span class="c">#: *NamedDict of sequence weights*, weights of sequences used to calculated scores</span>

    <span class="c"># if necessary calculate distance matrix</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">distanceMatrix</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Distance matrix not provided, calculating...&quot;</span>
        <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">calculateDistanceMatrix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Sequence weights not provided, calculating...&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">weights</span> <span class="o">=</span> <span class="n">calculateWeights</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span> <span class="n">distanceMatrix</span><span class="p">)</span>

    <span class="n">ConservationScores</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">records</span><span class="p">,</span><span class="s">&quot;Valdar score&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="o">.</span><span class="n">name</span><span class="p">,</span><span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">allocateScores</span><span class="p">()</span>
    <span class="n">startTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">valdarBySite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span><span class="p">,</span> <span class="n">weights</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="p">)</span>
    <span class="n">endTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Finished calculating scores for all sites </span><span class="si">%.3f</span><span class="s"> secs.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">endTime</span><span class="o">-</span><span class="n">startTime</span><span class="p">)</span>

  <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="ValdarScore.valdarBySite"><a class="viewcode-back" href="../index.html#profscore.ValdarScore.valdarBySite">[docs]</a>  <span class="k">def</span> <span class="nf">valdarBySite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">subsMat</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
      <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">      Calculate the Valdar conservation score for a given site in an alignment, given a list of sequence</span>
<span class="sd">      records, a substitution matrix, a site number and a distance matrix or list of sequence weights.</span>
<span class="sd">      Static function allows it to be called without creating a ``ValdarScore`` object.</span>
<span class="sd">      As per Valdar, 2001 PMID: 11093265</span>

<span class="sd">      Arguments:</span>

<span class="sd">        site : number</span>
<span class="sd">          site in alignment to calculate score for (note 0-based index)</span>
<span class="sd">        records : list of Bio.SeqRecord objects</span>
<span class="sd">          list of aligned sequences to process</span>
<span class="sd">        subsMat : Bio.SubsMat.SeqMat object</span>
<span class="sd">          substitution matrix, transformed</span>
<span class="sd">        weights : NamedDict of sequence weights</span>
<span class="sd">          pre-calculated weights of all sequences</span>
<span class="sd">        silent : boolean, optional</span>
<span class="sd">          if ``True`` do not produce any output (used when called by randomization functions), defaults to ``False``</span>

<span class="sd">      Returns:</span>

<span class="sd">        out : number</span>
<span class="sd">          Valdar score for the site.</span>
<span class="sd">      &quot;&quot;&quot;</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">weights</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Sequence weights must be provided&quot;</span><span class="p">)</span>

      <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Empty sequence records list provided!&quot;</span><span class="p">)</span>


      <span class="n">normaliser</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span> <span class="c"># normaliser (denominator) = sum of weights</span>
      <span class="n">score</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
      <span class="c"># do weighted sum of scores</span>
      <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
          <span class="k">for</span> <span class="n">recordJ</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
              <span class="k">if</span> <span class="n">recordJ</span><span class="o">&gt;</span><span class="n">recordI</span><span class="p">:</span>
                  <span class="n">seqI</span> <span class="o">=</span> <span class="n">recordI</span><span class="o">.</span><span class="n">seq</span>
                  <span class="n">seqJ</span> <span class="o">=</span> <span class="n">recordJ</span><span class="o">.</span><span class="n">seq</span>
                  <span class="n">weightI</span> <span class="o">=</span> <span class="n">weights</span><span class="p">[</span><span class="n">recordI</span><span class="p">]</span>
                  <span class="n">weightJ</span> <span class="o">=</span> <span class="n">weights</span><span class="p">[</span><span class="n">recordJ</span><span class="p">]</span>
                  <span class="n">weightIweightJ</span> <span class="o">=</span> <span class="n">weightI</span> <span class="o">*</span> <span class="n">weightJ</span>
                  <span class="n">normaliser</span> <span class="o">+=</span> <span class="n">weightIweightJ</span>

                  <span class="n">sj</span> <span class="o">=</span> <span class="n">seqI</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
                  <span class="n">sk</span> <span class="o">=</span> <span class="n">seqJ</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
                  <span class="n">Mut</span> <span class="o">=</span> <span class="n">subsMat</span><span class="p">[</span><span class="n">sj</span><span class="p">,</span><span class="n">sk</span><span class="p">]</span>
                  <span class="n">score</span><span class="o">+=</span><span class="n">weightIweightJ</span><span class="o">*</span><span class="n">Mut</span>

      <span class="c"># renormalise by sum of weights</span>
      <span class="n">normscore</span> <span class="o">=</span> <span class="n">score</span><span class="o">/</span><span class="n">normaliser</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">silent</span> <span class="ow">and</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;site </span><span class="si">%i</span><span class="s">: score </span><span class="si">%f</span><span class="s"> = </span><span class="si">%f</span><span class="s">/</span><span class="si">%f</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">normscore</span><span class="p">,</span><span class="n">score</span><span class="p">,</span><span class="n">normaliser</span><span class="p">)</span>
      <span class="k">return</span> <span class="n">normscore</span>


<span class="c">##########################################################################################</span>
<span class="c"># SEQUENCE PROFILE GENERATION AND SCORING</span>
<span class="c">##########################################################################################</span>
</div></div>
<div class="viewcode-block" id="AAScore"><a class="viewcode-back" href="../index.html#profscore.AAScore">[docs]</a><span class="k">class</span> <span class="nc">AAScore</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Simple data storage class to store and order AA scores for finding consensus.</span>
<span class="sd">    Sorting ``AAScore`` objects results in ordering by score.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">letter</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">pMax</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">letter</span> <span class="o">=</span> <span class="n">letter</span> <span class="c">#: *character*, letter for this aa</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span> <span class="c">#: *number*, absolute score for this aa</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pMax</span> <span class="o">=</span> <span class="n">pMax</span> <span class="c">#: *number*, score for this aa as a proportion of the max score at the site</span>
    <span class="k">def</span> <span class="nf">__cmp__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">cmp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">,</span><span class="n">other</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s">&#39;</span><span class="si">%s</span><span class="s">:</span><span class="si">%f</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">letter</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="SequenceProfile"><a class="viewcode-back" href="../index.html#profscore.SequenceProfile">[docs]</a><span class="k">class</span> <span class="nc">SequenceProfile</span><span class="p">():</span>
    <span class="sd">r&quot;&quot;&quot;</span>
<span class="sd">    Generate a SequenceProfile from set of alignment records.</span>
<span class="sd">    The sequence profile is the weighted average vector of amino acid scores for that position.</span>

<span class="sd">    The profile-based scoring system score breaks Valdar score into two parts, sequence profile generation (``SequenceProfile``) and profile scoring (``ProfileScore``).</span>

<span class="sd">    Profile automatically calculated when the class in constructed. Once constructed,</span>
<span class="sd">    this class will contain list (for each site in the alignment) of profile vectors as ``self.profile``.</span>

<span class="sd">    Sequence profile:</span>

<span class="sd">    .. math::</span>
<span class="sd">      \bar{P}_x = \frac{\displaystyle \sum_{i}^{N}w_iP_{ix}}{\displaystyle \sum_{i}^{N}w_i}</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">      vectorSubsMat : VectorSubsMat object</span>
<span class="sd">        substitution matrix, transformed and converted to vector form</span>
<span class="sd">      distanceMatrix : NamedDict of evolutionary distances, optional, may be ``None``</span>
<span class="sd">        pre-calculated distance matrix giving pairwise distances between all records (if not provided, calculated automatically; not used if profileWeights provided)</span>
<span class="sd">      profileWeights : NamedDict of sequence weights, optional, may be ``None``</span>
<span class="sd">        pre-calculated weights of all sequences used for weighting profile (if not provided, calculated automatically from evolutionary distances),</span>
<span class="sd">        weights for sequences used should sum to 1.0</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">vectorSubsMat</span><span class="p">,</span> <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">profileWeights</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Generating sequence profile...&#39;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">distanceMatrix</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">profileWeights</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Either distanceMatrix or profileWeights must be provided&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Empty sequence records list provided!&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">records</span> <span class="o">=</span> <span class="n">records</span> <span class="c">#: *list of Bio.SeqRecord objects*, list of records used to generate profile</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span> <span class="o">=</span> <span class="n">vectorSubsMat</span> <span class="c">#: *VectorSubsMat object*, vector substitution matrix used to generate profile</span>

        <span class="k">if</span> <span class="n">profileWeights</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">profileWeights</span> <span class="o">=</span> <span class="n">profileWeights</span> <span class="c">#: *NamedDict of sequence weights*, weights of sequences</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c">#calculate profileWeights, normalised to total of 1.0</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">profileWeights</span> <span class="o">=</span> <span class="n">calculateWeights</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span> <span class="n">distanceMatrix</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">checkTotalWeight</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">profileWeights</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;Total weight should be 1.0&#39;</span><span class="p">)</span>


        <span class="bp">self</span><span class="o">.</span><span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>  <span class="c">#: *number*, alignment length, i.e. total number of sites in alignment</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">consensus</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">          *list (for each site) of list (for multiple consensus residues) of AAScore objects*, ``None`` prior to call of ``findConsensus()``;</span>
<span class="sd">          stores consensus residues, as ``AAScore`` objects for each site, i.e. ``self.consensus[site]`` gives list of ``AAScore`` objects as consensus residues for the site</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">consensusDetails</span> <span class="o">=</span> <span class="bp">None</span> <span class="c">#: *string*, ``None`` prior to call of ``findConsensus()``;  details of how consensus determined, i.e. string giving cutoff values</span>

        <span class="c"># generate full profile</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">profile</span> <span class="o">=</span> <span class="p">[]</span> <span class="c">#: *list of numpy.array*, list of vectors for the profile, indexed by 0-based site number in alignment</span>

        <span class="n">sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span> <span class="c">#: *OrderedDict*, metadata for this profile</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;profile name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;Sequence profile&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;profile sequences&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%i</span><span class="s"> sequences (</span><span class="si">%08X</span><span class="s"> checksum)&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">),</span><span class="n">createRecordChecksum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;profile sequence ids&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;profile number of sites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%i</span><span class="s"> sites&quot;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;profile weighting&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">profileWeights</span><span class="o">.</span><span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;profile substitution matrix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">mat_name</span>
        
        <span class="n">startTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">):</span>
            <span class="n">profile_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generateProfileAtSite</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">profileWeights</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">profile</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">profile_x</span><span class="p">)</span>
        <span class="n">endTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Finished calculating profile for all sites </span><span class="si">%.3f</span><span class="s"> secs.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">endTime</span><span class="o">-</span><span class="n">startTime</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="SequenceProfile.generateProfileAtSite"><a class="viewcode-back" href="../index.html#profscore.SequenceProfile.generateProfileAtSite">[docs]</a>    <span class="k">def</span> <span class="nf">generateProfileAtSite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">vectorSubsMat</span><span class="p">,</span> <span class="n">profileWeights</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate the sequence profile at a site in the alignment.</span>
<span class="sd">        The sequence profile is the weighted average vector of amino acid scores for that position.</span>
<span class="sd">        Static function allows it to be called without creating a ``SequenceProfile`` object.</span>

<span class="sd">        Arguments:</span>

<span class="sd">          site : number</span>
<span class="sd">            position in the alignment to calculate profile (note: 0-based)</span>
<span class="sd">          records : list of Bio.SeqRecord objects</span>
<span class="sd">            list of aligned sequences to process</span>
<span class="sd">          vectorSubsMat : VectorSubsMat object</span>
<span class="sd">            substitution matrix, transformed and converted to vector form</span>
<span class="sd">          profileWeights : NamedDict of sequence weights</span>
<span class="sd">            pre-calculated weights of all sequences used for weighting profile (weights for sequences used assumed to sum to 1.0)</span>
<span class="sd">          silent : boolean, optional</span>
<span class="sd">            if ``True`` do not produce any output (used when called by randomization functions), defaults to ``False``</span>

<span class="sd">        Returns:</span>

<span class="sd">          out : numpy.array</span>
<span class="sd">            a vector giving the sequence profile at requested site</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">profile_x</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">dimensions</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">float64</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
            <span class="n">si</span> <span class="o">=</span> <span class="n">recordI</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
            <span class="n">weight</span> <span class="o">=</span> <span class="n">profileWeights</span><span class="p">[</span><span class="n">recordI</span><span class="p">]</span>
            <span class="n">profile_x</span> <span class="o">=</span> <span class="n">profile_x</span> <span class="o">+</span> <span class="n">weight</span><span class="o">*</span><span class="n">vectorSubsMat</span><span class="p">[</span><span class="n">si</span><span class="p">]</span> <span class="c">#add vector of scores to profile vector</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">silent</span> <span class="ow">and</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;site </span><span class="si">%i</span><span class="s">: profile </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s">&#39;</span><span class="si">%s</span><span class="s"> </span><span class="si">%f</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span><span class="p">[</span><span class="n">k</span><span class="p">],</span><span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">profile_x</span><span class="p">)]))</span>
        <span class="k">return</span> <span class="n">profile_x</span>
</div>
<div class="viewcode-block" id="SequenceProfile.findConsensus"><a class="viewcode-back" href="../index.html#profscore.SequenceProfile.findConsensus">[docs]</a>    <span class="k">def</span> <span class="nf">findConsensus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cutoffAbs</span><span class="p">,</span> <span class="n">cutoffProportion</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">excludeGap</span> <span class="o">=</span> <span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find consensus residues for each position in profile.</span>
<span class="sd">        For a residue to be a consensus residue at a site, it must be represented at that site in the alignment</span>
<span class="sd">        and must have a score at least equal to the cutoffs.</span>
<span class="sd">        Result is stored in attribute ``self.consensus``, with criteria given in ``self.consensusDetails``.</span>

<span class="sd">        Arguments:</span>

<span class="sd">          cutoffAbs : number</span>
<span class="sd">            absolute cutoff, residues must have absolutes scores &gt;= this to be considered a consensus residue</span>
<span class="sd">          cutoffProportion : number, optional</span>
<span class="sd">            proportional cutoff, residues must have scores &gt;= this proportion of the max score to be considered a consensus residue</span>
<span class="sd">          excludeGap : boolean, optional</span>
<span class="sd">            if ``True`` (default) prevents gaps from being considered consensus residues</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">consensus</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">consensusDetails</span> <span class="o">=</span> <span class="s">&#39;Consensus: residues with score &gt;= </span><span class="si">%.3f</span><span class="s"> or &gt;= </span><span class="si">%.2f</span><span class="s"> of max score&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">cutoffAbs</span><span class="p">,</span> <span class="n">cutoffProportion</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="bp">self</span><span class="o">.</span><span class="n">consensusDetails</span>
        <span class="n">letters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span>
        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">):</span>
            <span class="n">scoresAtSite</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">profile</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
            <span class="n">consensusAtSite</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">maxScore</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">scoresAtSite</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">maxScore</span><span class="o">&gt;=</span><span class="n">cutoffAbs</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="n">letters</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">excludeGap</span> <span class="ow">and</span> <span class="n">letter</span><span class="o">==</span><span class="s">&#39;-&#39;</span><span class="p">:</span> <span class="k">continue</span>
                    <span class="n">aaPresent</span> <span class="o">=</span> <span class="bp">False</span>
                    <span class="c"># filter residues by presence in alignment</span>
                    <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">:</span>
                        <span class="n">si</span> <span class="o">=</span> <span class="n">recordI</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">si</span><span class="o">==</span><span class="n">letter</span><span class="p">:</span>
                            <span class="n">aaPresent</span> <span class="o">=</span> <span class="bp">True</span>
                            <span class="k">break</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">aaPresent</span><span class="p">:</span> <span class="k">continue</span>
                    <span class="n">letterIndex</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">alphabetLUT</span><span class="p">[</span><span class="nb">ord</span><span class="p">(</span><span class="n">letter</span><span class="p">)]</span>
                    <span class="n">score</span> <span class="o">=</span> <span class="n">scoresAtSite</span><span class="p">[</span><span class="n">letterIndex</span><span class="p">]</span>
                    <span class="n">pMax</span> <span class="o">=</span> <span class="n">score</span><span class="o">/</span><span class="n">maxScore</span>
                    <span class="k">if</span> <span class="n">score</span><span class="o">&gt;=</span><span class="n">cutoffAbs</span> <span class="ow">or</span> <span class="n">pMax</span><span class="o">&gt;=</span><span class="n">cutoffProportion</span><span class="p">:</span>
                        <span class="n">aaScore</span> <span class="o">=</span> <span class="n">AAScore</span><span class="p">(</span><span class="n">letter</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">pMax</span><span class="p">)</span>
                        <span class="n">consensusAtSite</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aaScore</span><span class="p">)</span>
            <span class="n">consensusAtSite</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">consensus</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">consensusAtSite</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">3</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;site </span><span class="si">%i</span><span class="s">: consensus </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="s">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s"> </span><span class="si">%.3f</span><span class="s"> (</span><span class="si">%.2f</span><span class="s">)&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">aaScore</span><span class="o">.</span><span class="n">letter</span><span class="p">,</span><span class="n">aaScore</span><span class="o">.</span><span class="n">score</span><span class="p">,</span><span class="n">aaScore</span><span class="o">.</span><span class="n">pMax</span><span class="p">)</span> <span class="k">for</span> <span class="n">aaScore</span> <span class="ow">in</span> <span class="n">consensusAtSite</span><span class="p">))</span>
</div>
    <span class="k">def</span> <span class="nf">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">profile</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>

</div>
<div class="viewcode-block" id="ProfileScore"><a class="viewcode-back" href="../index.html#profscore.ProfileScore">[docs]</a><span class="k">class</span> <span class="nc">ProfileScore</span><span class="p">(</span><span class="n">ConservationScores</span><span class="p">):</span>
  <span class="sd">r&quot;&quot;&quot;</span>
<span class="sd">  Calculate sequence profile-based conservation scores for a set of alignment records given a sequence profile.</span>

<span class="sd">  The profile-based scoring system score breaks Valdar score into two parts, sequence profile generation (``SequenceProfile``) and profile scoring (``ProfileScore``).</span>
<span class="sd">  The score will be identical to the Valdar score if corrected for self-comparisons. This code does not assume that the sequence profile and profile scoring are</span>
<span class="sd">  generated from the same set of sequences, although it has not been tested or validated when different sets of sequences are used for each step.</span>

<span class="sd">  Scores are automatically calculated when the class in constructed. Once constructed,</span>
<span class="sd">  this class will contain list of conservation scores for each site in the alignment in ``self.scores``.</span>

<span class="sd">  Inherits from ``ConservationScores``.</span>

<span class="sd">  Uncorrected score:</span>

<span class="sd">  .. math ::</span>
<span class="sd">    \widetilde{C_x} = \frac{\displaystyle \sum_{i}^{N}w_i\bar{P}_{x}(s_{ix})}{\displaystyle \sum_{i}^{N}w_i}</span>

<span class="sd">  Corrected score:</span>

<span class="sd">  .. math ::</span>
<span class="sd">      C_x = \frac{\displaystyle \sum_{i}^{N}w_i\bar{P}_{x}(s_{ix}) -   \displaystyle \sum_{i}^{N}w_i w_i M\Big (s_{ix},s_{ix}\Big ) } {\displaystyle \sum_{i}^{N}w_i - \displaystyle \sum_{i}^{N}w_i w_i }</span>


<span class="sd">  Arguments:</span>

<span class="sd">    records : list of Bio.SeqRecord objects</span>
<span class="sd">      list of aligned sequences to process</span>
<span class="sd">    sequenceProfile : SequenceProfile object</span>
<span class="sd">      calculated sequence profile</span>
<span class="sd">    distanceMatrix : NamedDict of evolutionary distances, optional, may be ``None``</span>
<span class="sd">      pre-calculated distance matrix giving pairwise distances between all records (if not provided, calculated automatically; not used if weights provided)</span>
<span class="sd">    scoringWeights : NamedDict of sequence weights, optional, may be ``None``</span>
<span class="sd">      pre-calculated weights of sequences to be used for weighting the score (if not provided, calculated automatically from evolutionary distances),</span>
<span class="sd">      weights for sequences used should sum to 1.0</span>
<span class="sd">    correctForSelfComparisons : boolean, optional</span>
<span class="sd">      defaults to ``True``, if ``True``, score is corrected for self-comparisons</span>

<span class="sd">  &quot;&quot;&quot;</span>
  <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">sequenceProfile</span><span class="p">,</span> <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">scoringWeights</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">correctForSelfComparisons</span> <span class="o">=</span> <span class="bp">True</span><span class="p">):</span>

    <span class="n">correctedStr</span> <span class="o">=</span> <span class="s">&quot;corrected&quot;</span> <span class="k">if</span> <span class="n">correctForSelfComparisons</span> <span class="k">else</span> <span class="s">&quot;uncorrected&quot;</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Generating </span><span class="si">%s</span><span class="s"> profile-based conservation scores...&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">correctedStr</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">distanceMatrix</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">scoringWeights</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Either distanceMatrix or weights must be provided&quot;</span><span class="p">)</span>
    <span class="c">#if len(records)==0: raise ValueError(&quot;Empty sequence records list provided!&quot;)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">sequenceProfile</span> <span class="o">=</span> <span class="n">sequenceProfile</span> <span class="c">#: *SequenceProfile object*, sequenceProfile used to calculate scores</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">scoringWeights</span> <span class="o">=</span> <span class="bp">None</span> <span class="c">#: *NamedDict of sequence weights*, weights of sequences used to calculated scores</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">scoringWeights</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scoringWeights</span> <span class="o">=</span> <span class="n">calculateWeights</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">distanceMatrix</span><span class="p">)</span>  <span class="c">#: *NamedDict of sequence weights*, weights of sequences</span>
    <span class="k">else</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoringWeights</span> <span class="o">=</span> <span class="n">scoringWeights</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">checkTotalWeight</span><span class="p">(</span><span class="n">records</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">scoringWeights</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;Total weight should be 1.0&#39;</span><span class="p">)</span>

    <span class="n">ConservationScores</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">records</span><span class="p">,</span><span class="s">&quot;Profile-based score (</span><span class="si">%s</span><span class="s"> for self-comparisons)&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">correctedStr</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">scoringWeights</span><span class="o">.</span><span class="n">name</span><span class="p">,</span><span class="n">sequenceProfile</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s">&#39;profile name&#39;</span><span class="p">])</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">allocateScores</span><span class="p">()</span>
    <span class="n">vectorSubsMat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequenceProfile</span><span class="o">.</span><span class="n">vectorSubsMat</span>
    <span class="n">profileWeights</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequenceProfile</span><span class="o">.</span><span class="n">profileWeights</span>
    
    <span class="n">startTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">):</span>
        <span class="n">profileAtSite</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequenceProfile</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="n">ProfileScore</span><span class="o">.</span><span class="n">scoreAgainstProfileBySite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span> <span class="n">profileAtSite</span><span class="p">,</span> <span class="n">vectorSubsMat</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoringWeights</span><span class="p">,</span> <span class="n">profileWeights</span><span class="p">,</span> <span class="n">correctForSelfComparisons</span> <span class="o">=</span> <span class="n">correctForSelfComparisons</span><span class="p">)</span>
    <span class="n">endTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Finished calculating profile scores for all sites </span><span class="si">%.3f</span><span class="s"> secs.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">endTime</span><span class="o">-</span><span class="n">startTime</span><span class="p">)</span>


  <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="ProfileScore.scoreAgainstProfileBySite"><a class="viewcode-back" href="../index.html#profscore.ProfileScore.scoreAgainstProfileBySite">[docs]</a>  <span class="k">def</span> <span class="nf">scoreAgainstProfileBySite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">profileAtSite</span><span class="p">,</span> <span class="n">vectorSubsMat</span><span class="p">,</span> <span class="n">scoringWeights</span><span class="p">,</span> <span class="n">profileWeights</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">correctForSelfComparisons</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate conservation score at a site in the alignment.</span>
<span class="sd">    The score is based on comparison to ``sequenceProfile``.</span>
<span class="sd">    Static function allows it to be called without creating a ``ProfileScore`` object.</span>

<span class="sd">    Arguments:</span>

<span class="sd">        site : number</span>
<span class="sd">            position in the alignment to calculate profile (note: 0-based)</span>
<span class="sd">        records : list of Bio.SeqRecord objects</span>
<span class="sd">            list of aligned sequences to process</span>
<span class="sd">        profileAtSite : k-dimensional numpy array</span>
<span class="sd">            sequence profile at the site i.e weighted average vector of amino acid scores at site</span>
<span class="sd">        vectorSubsMat : VectorSubsMat object</span>
<span class="sd">            substitution matrix, transformed and converted to vector form; used for self-comparison correction</span>
<span class="sd">        scoringWeights : NamedDict of sequence weights</span>
<span class="sd">            pre-calculated weights of sequences to be used for weighting the score, weights for sequences used assumed to sum to 1.0</span>
<span class="sd">        profileWeights : NamedDict of sequence weights, optional</span>
<span class="sd">            pre-calculated weights of sequences to be used for weighting the profile; if absent assumed to be the same as sequence weights</span>
<span class="sd">        correctForSelfComparisons : boolean, optional</span>
<span class="sd">            defaults to ``True``, if ``True``, score is corrected for self-comparisons</span>
<span class="sd">        silent : boolean, optional</span>
<span class="sd">            if ``True`` do not produce any output (used when called by randomization functions), defaults to ``False``</span>

<span class="sd">    Returns:</span>

<span class="sd">      out : number</span>
<span class="sd">        conservation score for this site</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">mutAll</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span> <span class="c"># summed mut score for all sequences (including self-comparisons)</span>
    <span class="n">weightAll</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span> <span class="c"># summed weights for all sequences (including self-comparisons), assumed to be 1.0</span>
    <span class="n">mutDiagonal</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span> <span class="c"># summed mut score for self-comparisons</span>
    <span class="n">weightDiagonal</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span> <span class="c"># summed weights for self-comparisons</span>

    <span class="k">if</span> <span class="n">profileWeights</span> <span class="ow">is</span> <span class="bp">None</span> <span class="p">:</span> <span class="n">profileWeights</span> <span class="o">=</span> <span class="n">scoringWeights</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">correctForSelfComparisons</span><span class="p">:</span> <span class="n">profileWeights</span> <span class="o">=</span> <span class="p">[]</span> <span class="c"># cheat to skip self-comparison calculation block in loop</span>

    <span class="c"># Two kinds of sequence weight used:</span>
    <span class="c"># scoringWeights</span>
    <span class="c">#  = Weight of each sequence used when generating scores, i.e. contribution of a sequence to conservation score given profile</span>
    <span class="c"># profileWeight</span>
    <span class="c">#  = Weight of each sequence used when generating profile, i.e. contribution of a sequence to the sequence profile</span>
    <span class="c"># both stored so that the sum of weights for all sequences is 1.0</span>
    <span class="c">#</span>
    <span class="c"># As scoring scheme is designed for the same set of sequence set used for generating profile and subset, these two will be generally by identical.</span>
    <span class="c"># However, coded differently to allow for possibly use of different profile vs scoring sequences</span>

    <span class="k">for</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
      <span class="n">sij</span> <span class="o">=</span> <span class="n">recordI</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="c"># aa at site i in seq j</span>
      <span class="n">letterIndex</span> <span class="o">=</span> <span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">alphabetLUT</span><span class="p">[</span><span class="nb">ord</span><span class="p">(</span><span class="n">sij</span><span class="p">)]</span> <span class="c"># index of this aa in vector</span>

      <span class="n">weightI</span> <span class="o">=</span> <span class="n">scoringWeights</span><span class="p">[</span><span class="n">recordI</span><span class="p">]</span> <span class="c"># scoring weight of seq j</span>
      <span class="n">m_ij</span> <span class="o">=</span> <span class="n">profileAtSite</span><span class="p">[</span><span class="n">letterIndex</span><span class="p">]</span> <span class="c"># score of seq j from profile vector</span>

      <span class="k">if</span> <span class="n">recordI</span> <span class="ow">in</span> <span class="n">profileWeights</span><span class="p">:</span>
        <span class="c"># seq j was in profile, calculate self-specific mut needed for self-comparison comparison</span>
        <span class="n">weightP_I</span> <span class="o">=</span> <span class="n">profileWeights</span><span class="p">[</span><span class="n">recordI</span><span class="p">]</span> <span class="c"># profile weight of seq j</span>
        <span class="n">mutDiagonal</span> <span class="o">+=</span> <span class="n">weightI</span> <span class="o">*</span> <span class="n">weightP_I</span> <span class="o">*</span> <span class="n">vectorSubsMat</span><span class="p">[</span><span class="n">sij</span><span class="p">][</span><span class="n">letterIndex</span><span class="p">]</span>
        <span class="n">weightDiagonal</span> <span class="o">+=</span> <span class="n">weightI</span> <span class="o">*</span> <span class="n">weightP_I</span>  <span class="c"># sum all diagonal weights</span>

      <span class="n">mutAll</span><span class="o">+=</span><span class="n">weightI</span> <span class="o">*</span> <span class="n">m_ij</span>

    <span class="k">if</span> <span class="n">correctForSelfComparisons</span><span class="p">:</span>
      <span class="c"># corrected score = mutNonDiagonal/weightNonDiagonal = (mutAll - mutDiagonal) / (weightAll - weightNonDiagonal) = (mutAll - mutDiagonal) / (1.0 - weightNonDiagonal)</span>
      <span class="n">score</span> <span class="o">=</span> <span class="p">(</span><span class="n">mutAll</span> <span class="o">-</span> <span class="n">mutDiagonal</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">weightAll</span> <span class="o">-</span> <span class="n">weightDiagonal</span><span class="p">)</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">silent</span> <span class="ow">and</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;site </span><span class="si">%i</span><span class="s">: score </span><span class="si">%f</span><span class="s">, all </span><span class="si">%f</span><span class="s">/</span><span class="si">%f</span><span class="s">, diagonal </span><span class="si">%f</span><span class="s">/</span><span class="si">%f</span><span class="s">, non-diagonal </span><span class="si">%f</span><span class="s">/</span><span class="si">%f</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">score</span><span class="p">,</span><span class="n">mutAll</span><span class="p">,</span><span class="n">weightAll</span><span class="p">,</span><span class="n">mutDiagonal</span><span class="p">,</span><span class="n">weightDiagonal</span><span class="p">,</span><span class="n">mutAll</span> <span class="o">-</span> <span class="n">mutDiagonal</span><span class="p">,</span><span class="n">weightAll</span><span class="o">-</span><span class="n">weightDiagonal</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="c"># uncorrected score = mutAll/weightAll; weightAll is normalised to 1.0</span>
      <span class="n">score</span> <span class="o">=</span> <span class="n">mutAll</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">silent</span> <span class="ow">and</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;site </span><span class="si">%i</span><span class="s">: score </span><span class="si">%f</span><span class="s">, all </span><span class="si">%f</span><span class="s">/</span><span class="si">%f</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">mutAll</span><span class="p">,</span><span class="mf">1.0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">score</span>


<span class="c">##########################################################################################</span>
<span class="c"># SEQUENCE SAMPLING</span>
<span class="c">##########################################################################################</span>
</div></div>
<div class="viewcode-block" id="randomSampleReplacement"><a class="viewcode-back" href="../index.html#profscore.randomSampleReplacement">[docs]</a><span class="k">def</span> <span class="nf">randomSampleReplacement</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">number</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return a sample of sequences by sampling with replacement.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">      number : number</span>
<span class="sd">        number of sequences to sample</span>

<span class="sd">    Return:</span>

<span class="sd">      out : list of Bio.SeqRecord objects</span>
<span class="sd">        random sample of sequences</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sampleIndexes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">lastIndex</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">sampleIndexes</span><span class="p">)</span><span class="o">&lt;</span><span class="n">number</span><span class="p">:</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">lastIndex</span><span class="p">)</span>
        <span class="n">sampleIndexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">records</span><span class="p">[</span><span class="n">sampleIndex</span><span class="p">]</span> <span class="k">for</span> <span class="n">sampleIndex</span> <span class="ow">in</span> <span class="n">sampleIndexes</span><span class="p">]</span>
</div>
<div class="viewcode-block" id="randomSampleNoReplacement"><a class="viewcode-back" href="../index.html#profscore.randomSampleNoReplacement">[docs]</a><span class="k">def</span> <span class="nf">randomSampleNoReplacement</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">number</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return a sample of sequences by sampling without replacement.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">      number : number</span>
<span class="sd">        number of sequences to sample</span>

<span class="sd">    Return:</span>

<span class="sd">      out : list of Bio.SeqRecord objects</span>
<span class="sd">        random sample of sequences</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">number</span><span class="o">&gt;</span><span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">):</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Cannot sample more sequences than number provided!&quot;</span><span class="p">)</span>
    <span class="n">sampleIndexes</span> <span class="o">=</span> <span class="n">Set</span><span class="p">()</span>
    <span class="n">lastIndex</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">sampleIndexes</span><span class="p">)</span><span class="o">&lt;</span><span class="n">number</span><span class="p">:</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">lastIndex</span><span class="p">)</span>
        <span class="n">sampleIndexes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">records</span><span class="p">[</span><span class="n">sampleIndex</span><span class="p">]</span> <span class="k">for</span> <span class="n">sampleIndex</span> <span class="ow">in</span> <span class="n">sampleIndexes</span><span class="p">]</span>


<span class="c">##########################################################################################</span>
<span class="c"># SCORE PROBABILITY DISTRIBUTION</span>
<span class="c">##########################################################################################</span>
</div>
<div class="viewcode-block" id="AlignmentPDF"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF">[docs]</a><span class="k">class</span> <span class="nc">AlignmentPDF</span><span class="p">:</span>
    <span class="sd">r&quot;&quot;&quot;</span>
<span class="sd">    A conservation score probability distribution for the alignment. Generated by repeatedly randomly</span>
<span class="sd">    sampling a certain number of sequences from a larger set of sequences (without replacement).</span>

<span class="sd">    When constructed, this object will store and validate arguments provided in constructor.</span>
<span class="sd">    The object can then be used to:</span>

<span class="sd">      * Open an saved PDF and check for compatibility and consistency with this object</span>
<span class="sd">        using ``readFromFile(checkFile = True)``.</span>
<span class="sd">      * Calculate a PDF by performing randomization iterations using ``randomize()``, possibly</span>
<span class="sd">        appending existing iterations from an input file.</span>
<span class="sd">      * Open an saved PDF and look up densities site-by-site using ``prepareLookup()`` and</span>
<span class="sd">        ``lookupDensities()``.</span>

<span class="sd">    Arguments:</span>

<span class="sd">        records : list of Bio.SeqRecord objects</span>
<span class="sd">            list of aligned sequences to process</span>
<span class="sd">        sampleSize : number</span>
<span class="sd">            sample size for random sampling operation</span>
<span class="sd">        scoreType : ScoreType.VALDAR, ScoreType.PROFILE or ScoreType.PROFILE_UNCORR</span>
<span class="sd">            type of score to calculate calculate</span>
<span class="sd">        subsMat : Bio.SubsMat.SeqMat or VectorSubsMat object</span>
<span class="sd">            substitution matrix, transformed and converted to vector form if profile-based score being calculated</span>
<span class="sd">        distanceMatrix : NamedDict of evolutionary distances, optional, may be ``None``</span>
<span class="sd">            pre-calculated distance matrix giving pairwise distances between all records (if not provided, calculated automatically)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">startMagicWord</span> <span class="o">=</span> <span class="s">&#39;ProfileConsScore_PDFRand01_START&#39;</span> <span class="c">#: *string*, magic word at start of a saved ``ScoreProbDist``</span>
    <span class="n">endMagicWord</span> <span class="o">=</span> <span class="s">&#39;ProfileConsScore_PDFRand01_END&#39;</span> <span class="c">#: *string*, magic word at end of a saved ``ScoreProbDist``</span>
    <span class="n">dataMagicWord</span> <span class="o">=</span> <span class="s">&#39;note header ends&#39;</span> <span class="c">#: *string*, magic word for beginning of data section in file</span>
    <span class="n">sequenceIdHeader</span> <span class="o">=</span> <span class="s">&#39;sequence ids&#39;</span> <span class="c">#: *string*, header used for sequence ids</span>
    <span class="n">randomizationsHeader</span> <span class="o">=</span> <span class="s">&#39;total randomizations&#39;</span> <span class="c">#: *string*, header used for number of randomizations</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">sampleSize</span><span class="p">,</span> <span class="n">scoreType</span><span class="p">,</span> <span class="n">subsMat</span><span class="p">,</span> <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">records</span> <span class="o">=</span> <span class="n">records</span> <span class="c">#: *list of Bio.SeqRecord objects*, list of records provided to this score object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sampleSize</span> <span class="o">=</span> <span class="n">sampleSize</span> <span class="c">#: *number*, sample size of random sample</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">=</span> <span class="n">scoreType</span> <span class="c">#: *ScoreType*, type of score</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">distanceMatrix</span> <span class="c">#: *NamedDict of evolutionary distances*, matrix of distances between all sequences</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">subsMat</span><span class="p">,</span><span class="n">VectorSubsMat</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span> <span class="o">=</span> <span class="n">subsMat</span> <span class="c">#: *VectorSubsMat object*, vector substitution matrix to use for scoring (if profile-based scoring used)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">subsMat</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span> <span class="o">=</span> <span class="n">subsMat</span> <span class="c">#: *Bio.SubsMat.SeqMat object*, subsitution matrix to use for scoring (if Valdar scoring used)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">distanceMatrix</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">calculateDistanceMatrix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="c">#: *number*, alignment length, i.e. total number of sites in alignment</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#: *number*, total number of randomizations in this PDF</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">curSite</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#: *number*, current site being read or processed</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">csvReader</span> <span class="o">=</span> <span class="bp">None</span> <span class="c">#: *csv.CSVReader* object, stored for use when reading randomization file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">=</span> <span class="bp">None</span> <span class="c">#: one of *FilePosition* values, used when reading randomization file</span>

<div class="viewcode-block" id="AlignmentPDF.prepareMetadata"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF.prepareMetadata">[docs]</a>    <span class="k">def</span> <span class="nf">prepareMetadata</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate and return metadata for this object.</span>

<span class="sd">        Return:</span>

<span class="sd">          out : OrderedDict</span>
<span class="sd">            OrderedDict containing metadata for this object.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">metadata</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>


        <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;sequence number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">)</span>
        <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;sequence checksum&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%08X</span><span class="s">&quot;</span><span class="o">%</span><span class="n">createRecordChecksum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">)</span>
        <span class="n">metadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sequenceIdHeader</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">)</span>
        <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;sequence number of sites&#39;</span><span class="p">]</span> <span class="o">=</span>  <span class="s">&quot;</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span>

        <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;sample size&#39;</span><span class="p">]</span> <span class="o">=</span>  <span class="s">&quot;</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">sampleSize</span>
        <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;score type&#39;</span><span class="p">]</span> <span class="o">=</span>  <span class="n">ScoreType</span><span class="o">.</span><span class="n">reverse_mapping</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span><span class="p">]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="p">:</span>
            <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;substitution matrix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">mat_name</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;substitution matrix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span><span class="o">.</span><span class="n">mat_name</span>
        <span class="n">metadata</span><span class="p">[</span><span class="s">&#39;distance matrix type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">distanceMatrix</span><span class="o">.</span><span class="n">name</span>
        <span class="n">metadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">randomizationsHeader</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span>

        <span class="n">metadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">dataMagicWord</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;data follows immediately below&quot;</span>

        <span class="k">return</span> <span class="n">metadata</span>
</div>
<div class="viewcode-block" id="AlignmentPDF.randomize"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF.randomize">[docs]</a>    <span class="k">def</span> <span class="nf">randomize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outputHandle</span><span class="p">,</span> <span class="n">numRandomizations</span><span class="p">,</span> <span class="n">inputHandle</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform a certain number of randomizations to generate a PDF.</span>

<span class="sd">        In each randomization iteration, a random sample set of sequences is drawn</span>
<span class="sd">        for the list of records. Conservation scores are then calculated for this sample</span>
<span class="sd">        as per the method and options provided in the constructor.</span>

<span class="sd">        To conserve memory, randomizations are performed site-by-site. Scores for all</span>
<span class="sd">        randomizations for a given site are sorted in ascending order, binned (where</span>
<span class="sd">        identical scores are repeatedly obtained) and saved to disk.</span>

<span class="sd">        If an input file handle is given, randomizations will be</span>
<span class="sd">        read from that file and then appended to the new randomizations. The input file</span>
<span class="sd">        is checked for compatibility prior to starting.</span>

<span class="sd">        Once calculated, the data must be reread from the output file to lookup values</span>
<span class="sd">        from the PDF (see ``prepareLookup``).</span>

<span class="sd">        Arguments:</span>

<span class="sd">          outputHandle : output file handle</span>
<span class="sd">            handle for randomization output file used for writing randomizations;</span>
<span class="sd">            for each site, data written immediately after randomizations for that site performed.</span>
<span class="sd">            if ``None``, data is not written (for benchmarking)</span>
<span class="sd">          numRandomizations : number</span>
<span class="sd">            number of randomizations to perform</span>
<span class="sd">          inputHandle : input file handle, optional</span>
<span class="sd">            handle for randomization input file used for reading existing randomizations;</span>
<span class="sd">            these are combined with the (additional) randomizations performed here prior to</span>
<span class="sd">            writing; if ``None`` (default), no existing randomizations will be read</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">outputHandle</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
          <span class="n">csvWriter</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">outputHandle</span><span class="p">)</span>
          <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">startMagicWord</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">inputHandle</span><span class="p">:</span>
          <span class="bp">self</span><span class="o">.</span><span class="n">readFromFile</span><span class="p">(</span><span class="n">inputHandle</span><span class="p">)</span> <span class="c"># read any existing randomizations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span> <span class="o">+=</span> <span class="n">numRandomizations</span>
        <span class="n">reportMetadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">prepareMetadata</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Writing header...&#39;</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">reportMetadata</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
          <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;: &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">item</span><span class="p">])</span>
          <span class="k">if</span> <span class="n">outputHandle</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span> <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Performing randomizations using </span><span class="si">%s</span><span class="s"> score...&#39;</span><span class="o">%</span><span class="n">ScoreType</span><span class="o">.</span><span class="n">reverse_mapping</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39; Total </span><span class="si">%i</span><span class="s"> sequences, sampling </span><span class="si">%i</span><span class="s"> sequences.&#39;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">sampleSize</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39; Total </span><span class="si">%i</span><span class="s"> sites.&#39;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span>


        <span class="n">scores</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">numRandomizations</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">float64</span><span class="p">)</span>
        <span class="n">allSamples</span> <span class="o">=</span> <span class="p">[</span><span class="bp">None</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">numRandomizations</span><span class="p">)]</span>
        <span class="n">allWeights</span> <span class="o">=</span> <span class="p">[</span><span class="bp">None</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">numRandomizations</span><span class="p">)]</span>

        <span class="n">seed</span> <span class="o">=</span> <span class="nb">long</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">*</span> <span class="mi">256</span><span class="p">)</span>
        <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">randomization</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">numRandomizations</span><span class="p">):</span>
          <span class="c"># move these out of inner loops below to speed up running (at cost of more memory)</span>
          <span class="n">sample</span> <span class="o">=</span> <span class="n">randomSampleNoReplacement</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">sampleSize</span><span class="p">)</span>
          <span class="n">weights</span> <span class="o">=</span> <span class="n">calculateWeights</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">distanceMatrix</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
          <span class="n">allSamples</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span> <span class="o">=</span> <span class="n">sample</span>
          <span class="n">allWeights</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span> <span class="o">=</span> <span class="n">weights</span>

        <span class="n">rndStartTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">):</span>
            <span class="n">site1</span> <span class="o">=</span> <span class="n">site</span> <span class="o">+</span> <span class="mi">1</span> <span class="c"># 1-based index, used for output, internally 0-based used</span>
            <span class="n">scores</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">iterStartTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">VALDAR</span><span class="p">:</span>
                <span class="c"># randomize with Valdar scoring</span>
                <span class="k">for</span> <span class="n">randomization</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">numRandomizations</span><span class="p">):</span>
                    <span class="n">sample</span> <span class="o">=</span> <span class="n">allSamples</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span>
                    <span class="n">weights</span> <span class="o">=</span> <span class="n">allWeights</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span>
                    <span class="n">scores</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span> <span class="o">=</span> <span class="n">ValdarScore</span><span class="o">.</span><span class="n">valdarBySite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>

            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE_UNCORR</span><span class="p">:</span>
                <span class="n">correctForSelfComparisons</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">randomization</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">numRandomizations</span><span class="p">):</span>
                    <span class="n">sample</span> <span class="o">=</span> <span class="n">allSamples</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span>
                    <span class="n">weights</span> <span class="o">=</span> <span class="n">allWeights</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span>
                    <span class="n">profileAtSite</span> <span class="o">=</span> <span class="n">SequenceProfile</span><span class="o">.</span><span class="n">generateProfileAtSite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
                    <span class="n">scores</span><span class="p">[</span><span class="n">randomization</span><span class="p">]</span> <span class="o">=</span> <span class="n">ProfileScore</span><span class="o">.</span><span class="n">scoreAgainstProfileBySite</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">profileAtSite</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="n">correctForSelfComparisons</span> <span class="o">=</span> <span class="n">correctForSelfComparisons</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">AssertionError</span><span class="p">(</span><span class="s">&quot;Invalid score type.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span>
              <span class="k">print</span> <span class="s">&quot;  Random scores:&quot;</span>
              <span class="k">print</span> <span class="n">scores</span>

            <span class="n">sitePDF</span> <span class="o">=</span> <span class="n">SitePDF</span><span class="o">.</span><span class="n">createFromScores</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
            <span class="n">iterEndTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;  Completed randomizations for site </span><span class="si">%i</span><span class="s">, </span><span class="si">%.3f</span><span class="s"> secs.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">site1</span><span class="p">,</span><span class="n">iterEndTime</span><span class="o">-</span><span class="n">iterStartTime</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">inputHandle</span><span class="p">:</span>
              <span class="n">prevSitePDF</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">readNextSite</span><span class="p">()</span>
              <span class="n">sitePDF</span> <span class="o">=</span> <span class="n">SitePDF</span><span class="o">.</span><span class="n">createByCombining</span><span class="p">((</span><span class="n">prevSitePDF</span><span class="p">,</span><span class="n">sitePDF</span><span class="p">))</span>
              <span class="k">assert</span><span class="p">(</span><span class="n">site1</span><span class="o">==</span><span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="p">)</span>
            <span class="n">pointDensity</span> <span class="o">=</span> <span class="n">sitePDF</span><span class="o">.</span><span class="n">totalDensity</span><span class="p">()</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span><span class="o">!=</span><span class="n">pointDensity</span><span class="p">:</span>
              <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Unexpected number of randomizations, have </span><span class="si">%i</span><span class="s"> expect </span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">pointDensity</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">outputHandle</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span> <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">([</span><span class="n">site1</span><span class="p">],</span><span class="n">sitePDF</span><span class="o">.</span><span class="n">array</span><span class="p">)])</span>
        <span class="k">if</span> <span class="n">outputHandle</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
          <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">endMagicWord</span><span class="p">])</span>
          <span class="n">outputHandle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">inputHandle</span><span class="p">:</span> <span class="n">inputHandle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">rndEndTime</span> <span class="o">=</span> <span class="n">timeit</span><span class="o">.</span><span class="n">default_timer</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39; Completed all randomizations, </span><span class="si">%.3f</span><span class="s"> secs.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">rndEndTime</span><span class="o">-</span><span class="n">rndStartTime</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="AlignmentPDF.readFromFile"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF.readFromFile">[docs]</a>    <span class="k">def</span> <span class="nf">readFromFile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">inputHandle</span><span class="p">,</span> <span class="n">checkFile</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read a file containing a saved PDF, checking for compatibility with ``self``. Data in file can</span>
<span class="sd">        then be read to check for consistency (if ``checkFile = True``),</span>
<span class="sd">        or read site-by-site using ``readNextSite`` to obtain the PDF for each site.</span>

<span class="sd">        PDF given by ``inputHandle`` is tested for compatibility with arguments provided to constructor.</span>

<span class="sd">        Arguments:</span>

<span class="sd">          inputHandle : input file handle</span>
<span class="sd">            handle for randomization input file</span>
<span class="sd">          checkFile : boolean, optional</span>
<span class="sd">            if ``True``, entire file is read and checked for compatibility and consistency, if function returns</span>
<span class="sd">            without throwing exceptions then the file passes these checks;</span>
<span class="sd">            if ``False`` (default), header is read and checked for compatibility, and then reading stopped</span>
<span class="sd">            prior to reading data, site-by-site reading can then be carried out using ``readNextSite`` to</span>
<span class="sd">            return the PDF for each site</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">inputHandle</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="c"># reset file position</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csvReader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">inputHandle</span><span class="p">)</span>
        <span class="n">thisMetadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">prepareMetadata</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">=</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">START</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">csvReader</span><span class="p">):</span>
          <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span> <span class="k">continue</span> <span class="c"># skip blank lines</span>
          <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">==</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">START</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">startMagicWord</span><span class="p">:</span>
              <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file (line </span><span class="si">%i</span><span class="s">) does not start with magic word </span><span class="si">%s</span><span class="s"> - correct format?&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">index</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">startMagicWord</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">=</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">HEADER</span>
            <span class="n">fileMetadata</span> <span class="o">=</span> <span class="p">{}</span>
          <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">==</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">HEADER</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">)</span><span class="o">&lt;</span><span class="mi">2</span><span class="p">:</span>
              <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file header (line </span><span class="si">%i</span><span class="s">): Expected pairs of values - correct format?&quot;</span><span class="o">%</span><span class="n">index</span><span class="p">)</span>
            <span class="n">key</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">fileMetadata</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
            <span class="k">if</span> <span class="n">key</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataMagicWord</span><span class="p">:</span>
              <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">=</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">DATA</span>
              <span class="k">for</span> <span class="n">key</span><span class="p">,</span><span class="n">expectedValue</span> <span class="ow">in</span> <span class="n">thisMetadata</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">fileMetadata</span><span class="p">:</span>
                  <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file header (line </span><span class="si">%i</span><span class="s">): Value for key </span><span class="si">%s</span><span class="s"> does not exist&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">index</span><span class="p">,</span><span class="n">key</span><span class="p">))</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">fileMetadata</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">value</span><span class="o">!=</span><span class="n">expectedValue</span> <span class="ow">and</span> <span class="n">key</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">randomizationsHeader</span> <span class="ow">and</span> <span class="n">key</span><span class="o">!=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequenceIdHeader</span><span class="p">:</span>
                  <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file header (line </span><span class="si">%i</span><span class="s">): Value for key </span><span class="si">%s</span><span class="s"> does not match expected (</span><span class="si">%s</span><span class="s"> vs </span><span class="si">%s</span><span class="s">)&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">index</span><span class="p">,</span><span class="n">key</span><span class="p">,</span><span class="n">value</span><span class="p">,</span><span class="n">expectedValue</span><span class="p">))</span>
              <span class="bp">self</span><span class="o">.</span><span class="n">fileRandomizations</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fileMetadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">randomizationsHeader</span><span class="p">])</span> <span class="c"># store number of read randomizations</span>
              <span class="k">if</span> <span class="ow">not</span> <span class="n">checkFile</span><span class="p">:</span>
                <span class="c"># not checking file, i.e. reading to add to this</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fileRandomizations</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Randomization input file has </span><span class="si">%i</span><span class="s"> randomizations.&quot;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span>
              <span class="bp">self</span><span class="o">.</span><span class="n">curSite</span> <span class="o">=</span> <span class="mi">0</span> <span class="c"># 1-based, but incremented after successfully reading a site</span>
              <span class="k">break</span>
        <span class="k">if</span> <span class="n">checkFile</span><span class="p">:</span>
          <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">==</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">DATA</span><span class="p">:</span>
            <span class="k">while</span> <span class="bp">self</span><span class="o">.</span><span class="n">readNextSite</span><span class="p">(</span><span class="bp">True</span><span class="p">):</span>
              <span class="k">pass</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">!=</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">END</span><span class="p">:</span>
              <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file (line </span><span class="si">%i</span><span class="s">) does not end with magic word </span><span class="si">%s</span><span class="s">- correct format?&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">index</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">endMagicWord</span><span class="p">))</span>
</div>
<div class="viewcode-block" id="AlignmentPDF.readNextSite"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF.readNextSite">[docs]</a>    <span class="k">def</span> <span class="nf">readNextSite</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">checkFile</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read a data from input file site-by-site. Should be called after ``readFromFile`` to allow header</span>
<span class="sd">        to be read and ensure file pointer is in the correct position.</span>

<span class="sd">        Arguments:</span>

<span class="sd">          checkFile : boolean, optional</span>
<span class="sd">            if ``True``, returns boolean depending of whether site can be read correctly, used internally</span>
<span class="sd">            by ``readFromFile(checkFile = True)``;</span>
<span class="sd">            if ``False`` (default), read the next site and return ``SitePDF`` object for the site</span>

<span class="sd">        Return:</span>

<span class="sd">          out : SitePDF object or boolean</span>
<span class="sd">            if ``checkFile==False``, ``SitePDF`` for next site, position of site given by ``self.curSite``;</span>
<span class="sd">            if ``checkFile==True``, ``True`` if site can be read correctly</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">==</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">DATA</span><span class="p">:</span>
          <span class="k">while</span> <span class="bp">True</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">csvReader</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span> <span class="k">continue</span> <span class="c"># skip blank lines</span>
            <span class="k">else</span><span class="p">:</span> <span class="k">break</span>
          <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">endMagicWord</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="o">!=</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">:</span>
              <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file data: Reached end of file, last site </span><span class="si">%i</span><span class="s"> expected </span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filePosition</span> <span class="o">=</span> <span class="n">FilePosition</span><span class="o">.</span><span class="n">END</span>
            <span class="k">return</span> <span class="bp">None</span>
          <span class="k">try</span><span class="p">:</span>
            <span class="n">fileSite</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="o">+=</span><span class="mi">1</span> <span class="c"># increment after successfully reading a site</span>
            <span class="n">siteData</span> <span class="o">=</span> <span class="n">itertools</span><span class="o">.</span><span class="n">islice</span><span class="p">(</span><span class="n">row</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="bp">None</span><span class="p">)</span> <span class="c"># skip first element = site number</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="o">!=</span><span class="n">fileSite</span><span class="p">:</span>
              <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Got site </span><span class="si">%i</span><span class="s"> expected </span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">fileSite</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">checkFile</span><span class="p">:</span>
              <span class="n">SitePDF</span><span class="o">.</span><span class="n">convertFromStrList</span><span class="p">(</span><span class="n">siteData</span><span class="p">,</span><span class="bp">False</span><span class="p">)</span> <span class="c"># check data is correct format</span>
              <span class="k">return</span> <span class="bp">True</span>
            <span class="k">else</span><span class="p">:</span>
              <span class="n">sitePDF</span> <span class="o">=</span> <span class="n">SitePDF</span><span class="o">.</span><span class="n">convertFromStrList</span><span class="p">(</span><span class="n">siteData</span><span class="p">,</span><span class="bp">True</span><span class="p">)</span>
              <span class="n">pointDensity</span> <span class="o">=</span> <span class="n">sitePDF</span><span class="o">.</span><span class="n">totalDensity</span><span class="p">()</span>
              <span class="k">if</span> <span class="n">pointDensity</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fileRandomizations</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Read unexpected number of randomizations, read </span><span class="si">%i</span><span class="s"> expect </span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">pointDensity</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">fileRandomizations</span><span class="p">))</span>
              <span class="k">return</span> <span class="n">sitePDF</span>
          <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
              <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file data: Error parsing data </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

        <span class="k">else</span><span class="p">:</span>
          <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Randomization input file data: Unexpected file position&quot;</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="AlignmentPDF.combineFromMultiple"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF.combineFromMultiple">[docs]</a>    <span class="k">def</span> <span class="nf">combineFromMultiple</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outputHandle</span><span class="p">,</span> <span class="n">inputHandles</span><span class="p">):</span>
      <span class="n">csvWriter</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">outputHandle</span><span class="p">)</span>
      <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">startMagicWord</span><span class="p">])</span>

      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Combining randomizations from </span><span class="si">%i</span><span class="s"> files...&#39;</span><span class="o">%</span><span class="nb">len</span><span class="p">(</span><span class="n">inputHandles</span><span class="p">)</span>
      <span class="n">alignmentPDFs</span> <span class="o">=</span> <span class="p">[]</span>
      <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">inputHandle</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">inputHandles</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Checking file </span><span class="si">%i</span><span class="s"> for compatibility...&#39;</span><span class="o">%</span><span class="n">index</span>
        <span class="n">alignmentPDF</span> <span class="o">=</span> <span class="n">AlignmentPDF</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sampleSize</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoreType</span><span class="p">,</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectorSubsMat</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">subsMat</span><span class="p">,</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">distanceMatrix</span><span class="p">)</span>
        <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">readFromFile</span><span class="p">(</span><span class="n">inputHandle</span><span class="p">,</span> <span class="n">checkFile</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span> <span class="c"># check whole file</span>
        <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">readFromFile</span><span class="p">(</span><span class="n">inputHandle</span><span class="p">)</span> <span class="c"># reseek to data start position</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span> <span class="o">+=</span> <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">totalRandomizations</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Read </span><span class="si">%i</span><span class="s"> randomizations from file </span><span class="si">%i</span><span class="s">...&#39;</span><span class="o">%</span>\
                              <span class="p">(</span><span class="n">alignmentPDF</span><span class="o">.</span><span class="n">totalRandomizations</span><span class="p">,</span><span class="n">index</span><span class="p">)</span>
        <span class="n">alignmentPDFs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignmentPDF</span><span class="p">)</span>
      <span class="n">reportMetadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">prepareMetadata</span><span class="p">()</span>

      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Resulting file will have </span><span class="si">%i</span><span class="s"> randomizations...&#39;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span>

      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Writing header...&#39;</span>
      <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">reportMetadata</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;: &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">item</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">outputHandle</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span> <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>

      <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">):</span>
          <span class="n">site1</span> <span class="o">=</span> <span class="n">site</span> <span class="o">+</span> <span class="mi">1</span> <span class="c"># 1-based index, used for output, internally 0-based used</span>
          <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Combining site </span><span class="si">%i</span><span class="s">...&#39;</span><span class="o">%</span><span class="n">site1</span>
          <span class="n">sitePDFs</span> <span class="o">=</span> <span class="p">[]</span>
          <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">alignmentPDF</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignmentPDFs</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">3</span> <span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Reading site </span><span class="si">%i</span><span class="s"> from file </span><span class="si">%i</span><span class="s">...&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">site1</span><span class="p">,</span><span class="n">index</span><span class="p">)</span>
            <span class="n">sitePDF</span> <span class="o">=</span> <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">readNextSite</span><span class="p">()</span>
            <span class="k">assert</span><span class="p">(</span><span class="n">site1</span><span class="o">==</span><span class="n">alignmentPDF</span><span class="o">.</span><span class="n">curSite</span><span class="p">)</span>
            <span class="n">sitePDFs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sitePDF</span><span class="p">)</span>
          <span class="n">sitePDF</span> <span class="o">=</span> <span class="n">SitePDF</span><span class="o">.</span><span class="n">createByCombining</span><span class="p">(</span><span class="n">sitePDFs</span><span class="p">)</span>
          <span class="n">pointDensity</span> <span class="o">=</span> <span class="n">sitePDF</span><span class="o">.</span><span class="n">totalDensity</span><span class="p">()</span>
          <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span><span class="o">!=</span><span class="n">pointDensity</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Unexpected number of randomizations, have </span><span class="si">%i</span><span class="s"> expect </span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">pointDensity</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span><span class="p">))</span>
          <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">([</span><span class="n">site1</span><span class="p">],</span><span class="n">sitePDF</span><span class="o">.</span><span class="n">array</span><span class="p">)])</span>
      <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">endMagicWord</span><span class="p">])</span>
      <span class="n">outputHandle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
      <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">inputHandle</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">inputHandles</span><span class="p">):</span>
        <span class="n">inputHandle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39; Completed file combining.&#39;</span>
</div>
<div class="viewcode-block" id="AlignmentPDF.prepareLookup"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF.prepareLookup">[docs]</a>    <span class="k">def</span> <span class="nf">prepareLookup</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">inputHandle</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Prepare a file for looking-up values from the PDF site-by-site.</span>
<span class="sd">        PDF density for a site can then be found by calling ``lookupDensities``.</span>

<span class="sd">        Arguments:</span>

<span class="sd">          inputHandle : input file handle</span>
<span class="sd">            handle for randomization input file</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">readFromFile</span><span class="p">(</span><span class="n">inputHandle</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="AlignmentPDF.lookupDensities"><a class="viewcode-back" href="../index.html#profscore.AlignmentPDF.lookupDensities">[docs]</a>    <span class="k">def</span> <span class="nf">lookupDensities</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">site</span><span class="p">,</span> <span class="n">score</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Lookup densities in a prepared PDF. Must be called after ``prepareLookup()``.</span>

<span class="sd">        Returns density in the PDF where function values ≥ ``score``,</span>
<span class="sd">        as well as the total function density. As data is obtained by reading from current file position,</span>
<span class="sd">        must be called in increasing order of site numbers.</span>

<span class="sd">        Arguments:</span>

<span class="sd">          site : number</span>
<span class="sd">            index (0-based) of site to lookup density for, must be greater than ``self.curSite``</span>
<span class="sd">          score : number</span>
<span class="sd">            score to return density for</span>

<span class="sd">        Return:</span>

<span class="sd">          out : tuple of (number,number)</span>
<span class="sd">            (density where function ≥ ``score``, total density of function)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">site1</span> <span class="o">=</span> <span class="n">site</span><span class="o">+</span><span class="mi">1</span> <span class="c"># 1-based index, used in randomization file</span>
        <span class="k">while</span> <span class="bp">True</span><span class="p">:</span>
          <span class="c"># read correct site, may need to skip lines as not all sites looked up</span>
          <span class="n">sitePDF</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">readNextSite</span><span class="p">()</span>
          <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">3</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Read site </span><span class="si">%i</span><span class="s"> from file.&quot;</span><span class="o">%</span><span class="bp">self</span><span class="o">.</span><span class="n">curSite</span>
          <span class="k">assert</span><span class="p">(</span><span class="n">site1</span><span class="o">&gt;=</span><span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="p">)</span>
          <span class="k">if</span> <span class="n">site1</span><span class="o">==</span><span class="bp">self</span><span class="o">.</span><span class="n">curSite</span><span class="p">:</span> <span class="k">break</span>
        <span class="n">density</span> <span class="o">=</span> <span class="n">sitePDF</span><span class="o">.</span><span class="n">density</span><span class="p">(</span><span class="n">score</span><span class="p">,</span><span class="n">operator</span><span class="o">.</span><span class="n">ge</span><span class="p">)</span>
        <span class="n">totalDensity</span> <span class="o">=</span> <span class="n">sitePDF</span><span class="o">.</span><span class="n">totalDensity</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Looked up density for site </span><span class="si">%i</span><span class="s"> is </span><span class="si">%i</span><span class="s">/</span><span class="si">%i</span><span class="s"> when random score&gt;=</span><span class="si">%.4f</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site1</span><span class="p">,</span><span class="n">density</span><span class="p">,</span><span class="n">totalDensity</span><span class="p">,</span><span class="n">score</span><span class="p">)</span>
        <span class="k">assert</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">totalRandomizations</span><span class="o">==</span><span class="n">totalDensity</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">density</span><span class="p">,</span> <span class="n">totalDensity</span><span class="p">)</span>
</div></div>
<div class="viewcode-block" id="SitePDF"><a class="viewcode-back" href="../index.html#profscore.SitePDF">[docs]</a><span class="k">class</span> <span class="nc">SitePDF</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Probability density function for a single site (in an alignment).</span>
<span class="sd">    Stored as an array of discrete points (``PDFPoint`` objects), each containing a value and a density.</span>

<span class="sd">    Initialized as an empty PDF (no points), use class methods to create a ``PDFPoint``:</span>
<span class="sd">      * From a an array of scores, using ``createFromScores()``.</span>
<span class="sd">      * By converting from a list of strings (e.g. read from a csv file) using ``convertFromStrList()``.</span>
<span class="sd">      * By combining two existing ``PDFPoint`` objects, usng ``createByCombining()``</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s">&#39;array&#39;</span><span class="p">)</span> <span class="c"># use slots to save memory</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">array</span> <span class="o">=</span> <span class="p">[]</span> <span class="c">#: *list of PDFPoint objects*, array of points making up of this PDF</span>

    <span class="nd">@classmethod</span>
<div class="viewcode-block" id="SitePDF.createFromScores"><a class="viewcode-back" href="../index.html#profscore.SitePDF.createFromScores">[docs]</a>    <span class="k">def</span> <span class="nf">createFromScores</span><span class="p">(</span><span class="n">cls</span><span class="p">,</span> <span class="n">scores</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a new ``SitePDF`` object from an list/array of scores.</span>

<span class="sd">        Arguments:</span>

<span class="sd">            scores : list of numbers</span>
<span class="sd">                list of scores to convert</span>

<span class="sd">        Return:</span>

<span class="sd">            out : SitePDF object</span>
<span class="sd">                SitePDF containing points created from the supplied scores</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">this</span> <span class="o">=</span> <span class="n">cls</span><span class="p">()</span>

        <span class="c"># sort scores and the summarize by collapsing equal scores into single points</span>
        <span class="n">scores</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">curPDFPoint</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">score</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">scores</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">curPDFPoint</span> <span class="ow">and</span> <span class="n">curPDFPoint</span><span class="o">.</span><span class="n">approxEqual</span><span class="p">(</span><span class="n">score</span><span class="p">):</span>
                <span class="n">curPDFPoint</span><span class="o">.</span><span class="n">increment</span><span class="p">()</span> <span class="c"># same value, increment current point&#39;s density</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c"># different value, need new point</span>
                <span class="n">curPDFPoint</span> <span class="o">=</span> <span class="n">PDFPoint</span><span class="p">(</span><span class="n">score</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">this</span><span class="o">.</span><span class="n">array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curPDFPoint</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">this</span>
</div>
    <span class="nd">@classmethod</span>
<div class="viewcode-block" id="SitePDF.convertFromStrList"><a class="viewcode-back" href="../index.html#profscore.SitePDF.convertFromStrList">[docs]</a>    <span class="k">def</span> <span class="nf">convertFromStrList</span><span class="p">(</span><span class="n">cls</span><span class="p">,</span> <span class="n">strData</span><span class="p">,</span> <span class="n">store</span> <span class="o">=</span> <span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a new ``SitePDF`` object by converting from a list of strings to a series</span>
<span class="sd">        of points.</span>

<span class="sd">        Format is [``&#39;value1:density&#39;, &#39;value2:density2&#39;, ...]``,</span>
<span class="sd">        if ``:density`` is absent, density is assumed to be 1.</span>

<span class="sd">        This format for a point (``&#39;value(:density)&#39;``) is produced by ``str(pdfPoint)``.</span>
<span class="sd">        To produce a list of strings for all points in this format, use ``[str(item) for item in self.array]``.</span>

<span class="sd">        Arguments:</span>

<span class="sd">            strData : list of string</span>
<span class="sd">                string data to convert</span>
<span class="sd">            store : boolean, optional</span>
<span class="sd">                if ``True`` (default) read and store the data, otherwise just read without storing (for checking data format)</span>

<span class="sd">        Return:</span>

<span class="sd">            out : SitePDF object</span>
<span class="sd">                SitePDF containing points created from supplied data  or ``None`` if ``store`` is ``False``</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">store</span><span class="p">:</span> <span class="n">this</span> <span class="o">=</span> <span class="n">cls</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span> <span class="n">this</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">strData</span><span class="p">:</span>
            <span class="k">if</span> <span class="s">&#39;:&#39;</span> <span class="ow">in</span> <span class="n">element</span><span class="p">:</span>
              <span class="n">valueStr</span><span class="p">,</span> <span class="n">densityStr</span> <span class="o">=</span> <span class="n">element</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)</span>
              <span class="n">density</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">densityStr</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
              <span class="n">valueStr</span> <span class="o">=</span> <span class="n">element</span>
              <span class="n">density</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">float64</span><span class="p">(</span><span class="n">valueStr</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">store</span><span class="p">:</span>
              <span class="n">curPDFPoint</span> <span class="o">=</span> <span class="n">PDFPoint</span><span class="p">(</span><span class="n">value</span><span class="p">,</span><span class="n">density</span><span class="p">)</span>
              <span class="n">this</span><span class="o">.</span><span class="n">array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curPDFPoint</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">this</span>
</div>
    <span class="nd">@classmethod</span>
<div class="viewcode-block" id="SitePDF.createByCombining"><a class="viewcode-back" href="../index.html#profscore.SitePDF.createByCombining">[docs]</a>    <span class="k">def</span> <span class="nf">createByCombining</span><span class="p">(</span><span class="n">cls</span><span class="p">,</span> <span class="n">sitePDFs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a new ``SitePDF`` object by combining multiple ``SitePDF`` objects.</span>

<span class="sd">        Arguments:</span>

<span class="sd">            sitePDF : iterable of SitePDF objects</span>
<span class="sd">                objects to combine</span>

<span class="sd">        Return:</span>

<span class="sd">            out : SitePDF object</span>
<span class="sd">                SitePDF containing points created by combining the supplied objects</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">this</span> <span class="o">=</span> <span class="n">cls</span><span class="p">()</span>
        <span class="n">combinedArray</span> <span class="o">=</span> <span class="nb">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span><span class="o">+</span><span class="n">y</span><span class="p">,</span> <span class="p">(</span><span class="n">sitePDF</span><span class="o">.</span><span class="n">array</span> <span class="k">for</span> <span class="n">sitePDF</span> <span class="ow">in</span> <span class="n">sitePDFs</span><span class="p">))</span>
        <span class="n">combinedArray</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span> <span class="c"># combine then sort points</span>

        <span class="n">lastPoint</span> <span class="o">=</span> <span class="n">combinedArray</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">combinedArray</span><span class="p">)):</span> <span class="c"># iterate sorted points</span>
            <span class="n">point</span> <span class="o">=</span> <span class="n">combinedArray</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">point</span><span class="o">.</span><span class="n">combineIfEqual</span><span class="p">(</span><span class="n">lastPoint</span><span class="p">)</span> <span class="c"># check if equal value to previous point, if so, combine points leaving one point object with zero density</span>
            <span class="n">lastPoint</span> <span class="o">=</span> <span class="n">point</span>
        <span class="n">this</span><span class="o">.</span><span class="n">array</span> <span class="o">=</span> <span class="p">[</span><span class="n">point</span> <span class="k">for</span> <span class="n">point</span> <span class="ow">in</span> <span class="n">combinedArray</span> <span class="k">if</span> <span class="n">point</span><span class="o">.</span><span class="n">density</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">]</span> <span class="c"># create new array excluding any points with zero density</span>
        <span class="k">return</span> <span class="n">this</span>
</div>
<div class="viewcode-block" id="SitePDF.density"><a class="viewcode-back" href="../index.html#profscore.SitePDF.density">[docs]</a>    <span class="k">def</span> <span class="nf">density</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">oper</span><span class="p">):</span>
      <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">      Return summed density of this object for values that return ``True`` for ``oper(value,score)``.</span>
<span class="sd">      e.g. ``density(0.7,operator.ge)`` will return density where value is ≥ 0.7.</span>

<span class="sd">        Arguments:</span>

<span class="sd">            score : number</span>
<span class="sd">                comparison number</span>
<span class="sd">            oper : operator</span>
<span class="sd">                operator to use for comparison, e.g. ``operator.ge``</span>

<span class="sd">        Return:</span>

<span class="sd">          out : number</span>
<span class="sd">              summed density</span>
<span class="sd">      &quot;&quot;&quot;</span>
      <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">point</span><span class="o">.</span><span class="n">density</span> <span class="k">for</span> <span class="n">point</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">array</span> <span class="k">if</span> <span class="n">oper</span><span class="p">(</span><span class="n">point</span><span class="o">.</span><span class="n">value</span><span class="p">,</span><span class="n">score</span><span class="p">))</span>
</div>
<div class="viewcode-block" id="SitePDF.totalDensity"><a class="viewcode-back" href="../index.html#profscore.SitePDF.totalDensity">[docs]</a>    <span class="k">def</span> <span class="nf">totalDensity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
      <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">      Sum total density of this object.</span>

<span class="sd">        Return:</span>

<span class="sd">          out : number</span>
<span class="sd">              total density in this object</span>
<span class="sd">      &quot;&quot;&quot;</span>
      <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">point</span><span class="o">.</span><span class="n">density</span> <span class="k">for</span> <span class="n">point</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">array</span><span class="p">)</span>
</div></div>
<span class="nd">@total_ordering</span>
<div class="viewcode-block" id="PDFPoint"><a class="viewcode-back" href="../index.html#profscore.PDFPoint">[docs]</a><span class="k">class</span> <span class="nc">PDFPoint</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a single point of a probability density function containing a value and a density.</span>
<span class="sd">    Can be compared (``==, &lt;`` etc.) to numbers and other ``PDFPoint`` objects to allow ordering</span>
<span class="sd">    and testing whether this ``PDFPoint`` represents same value.</span>

<span class="sd">    Note that values are considered equal if their difference is less than a very small value to allow</span>
<span class="sd">    for discrete nature of IEEE floating point representation (see ``approxEqual()``).</span>

<span class="sd">    Allows conversion to string in the format ``value:density`` or ``value`` (if ``density=1``) using</span>
<span class="sd">    ``str(pdfPoint)``.</span>

<span class="sd">    Arguments:</span>

<span class="sd">        value : number</span>
<span class="sd">            value of the point</span>
<span class="sd">        density : number, optional</span>
<span class="sd">            density of the point, stored as an integer indicating</span>
<span class="sd">            number of times this value obtained by randomization</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s">&#39;value&#39;</span><span class="p">,</span><span class="s">&#39;density&#39;</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">density</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">value</span> <span class="o">=</span> <span class="n">value</span> <span class="c">#: *number*, value of this point</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">density</span> <span class="o">=</span> <span class="n">density</span> <span class="c">#: *number*, density of this point (integer)</span>

<div class="viewcode-block" id="PDFPoint.increment"><a class="viewcode-back" href="../index.html#profscore.PDFPoint.increment">[docs]</a>    <span class="k">def</span> <span class="nf">increment</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Increment the density of this point by one.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">density</span> <span class="o">+=</span> <span class="mi">1</span>
</div>
<div class="viewcode-block" id="PDFPoint.combineIfEqual"><a class="viewcode-back" href="../index.html#profscore.PDFPoint.combineIfEqual">[docs]</a>    <span class="k">def</span> <span class="nf">combineIfEqual</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Combine this ``PDFPoint`` with another if the two have an (approximately) equal value.</span>

<span class="sd">        Arguments:</span>

<span class="sd">            other : PDFPoint</span>
<span class="sd">                other ``PDFPoint`` to possibly combine with</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">approxEqual</span><span class="p">(</span><span class="n">other</span><span class="o">.</span><span class="n">value</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">density</span> <span class="o">+=</span> <span class="n">other</span><span class="o">.</span><span class="n">density</span>
            <span class="n">other</span><span class="o">.</span><span class="n">density</span> <span class="o">=</span> <span class="mi">0</span></div>
<div class="viewcode-block" id="PDFPoint.checkComparable"><a class="viewcode-back" href="../index.html#profscore.PDFPoint.checkComparable">[docs]</a>    <span class="k">def</span> <span class="nf">checkComparable</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">other</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check whether this ``PDFPoint`` can be compared to ``other``. Returns the value of</span>
<span class="sd">        ``other`` to compare against if ``other`` is another ``PDFPoint`` or same type as</span>
<span class="sd">        ``self.value``, returns ``None`` otherwise.</span>

<span class="sd">        Arguments:</span>

<span class="sd">            other : object</span>
<span class="sd">                other object to test</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">other</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">other</span><span class="o">.</span><span class="n">value</span>
        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">other</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">value</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">other</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">None</span></div>
<div class="viewcode-block" id="PDFPoint.approxEqual"><a class="viewcode-back" href="../index.html#profscore.PDFPoint.approxEqual">[docs]</a>    <span class="k">def</span> <span class="nf">approxEqual</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">otherValue</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return ``True`` if difference between ``self.value`` and ``otherValue`` is less than</span>
<span class="sd">        ``1e-12``, used to test equality given discrete nature of IEEE floating point</span>
<span class="sd">        representation and possible round-off error when calculating conservation scores.</span>

<span class="sd">        Arguments:</span>

<span class="sd">            otherValue : object</span>
<span class="sd">                other object to test</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">abs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">value</span><span class="o">-</span><span class="n">otherValue</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">1e-12</span></div>
    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">density</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
          <span class="k">return</span> <span class="s">&quot;</span><span class="si">%s</span><span class="s">:</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">value</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">density</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span> <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="n">otherValue</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">checkComparable</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">otherValue</span><span class="p">:</span> <span class="k">return</span> <span class="bp">False</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">approxEqual</span><span class="p">(</span><span class="n">otherValue</span><span class="p">)</span> <span class="c"># return true if other is approx equal</span>
    <span class="k">def</span> <span class="nf">__lt__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="n">otherValue</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">checkComparable</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">otherValue</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&#39;Comparing PDFPoint to unsupported type&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">value</span><span class="o">&lt;</span><span class="n">otherValue</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">approxEqual</span><span class="p">(</span><span class="n">other</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>


<span class="c">##########################################################################################</span>
<span class="c"># REPORT OUTPUT</span>
<span class="c">##########################################################################################</span>
</div>
<div class="viewcode-block" id="getAllSites"><a class="viewcode-back" href="../index.html#profscore.getAllSites">[docs]</a><span class="k">def</span> <span class="nf">getAllSites</span><span class="p">(</span><span class="n">records</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return a list of all sites in the alignment.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="c"># alignment length</span>
    <span class="n">allSites</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">allSites</span>
</div>
<div class="viewcode-block" id="getPresentSites"><a class="viewcode-back" href="../index.html#profscore.getPresentSites">[docs]</a><span class="k">def</span> <span class="nf">getPresentSites</span><span class="p">(</span><span class="n">records</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return a list of all present sites in the alignment.</span>
<span class="sd">    A present site must be represented by a non-gap character</span>
<span class="sd">    in at least one sequence.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      records : list of Bio.SeqRecord objects</span>
<span class="sd">        list of aligned sequences to process</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">presentSites</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="c"># alignment length</span>
    <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">):</span>
        <span class="n">sitePresent</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&#39;-&#39;</span><span class="p">:</span>
                <span class="n">sitePresent</span> <span class="o">=</span> <span class="bp">True</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="n">sitePresent</span><span class="p">:</span>
            <span class="n">presentSites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">presentSites</span>
</div>
<div class="viewcode-block" id="SequenceNumerator"><a class="viewcode-back" href="../index.html#profscore.SequenceNumerator">[docs]</a><span class="k">class</span> <span class="nc">SequenceNumerator</span><span class="p">:</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Iterator for sequence numbering.</span>
<span class="sd">  Skips from -1 to +1 (no 0).</span>
<span class="sd">  Allows subnumbering for gap positions, i.e. 1,1a,1b,1c.</span>

<span class="sd">  Arguments:</span>

<span class="sd">    start : number</span>
<span class="sd">      starting number</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">subnums</span> <span class="o">=</span> <span class="s">&#39;abcdefghijklmnopqrstuvwxyz&#39;</span> <span class="c">#: *string*, suffixes to use for subnumbering</span>
  <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="p">):</span>
      <span class="bp">self</span><span class="o">.</span><span class="n">reset</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>

  <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
      <span class="k">return</span> <span class="bp">self</span>

<div class="viewcode-block" id="SequenceNumerator.reset"><a class="viewcode-back" href="../index.html#profscore.SequenceNumerator.reset">[docs]</a>  <span class="k">def</span> <span class="nf">reset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">start</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reset numbering to new starting position.</span>

<span class="sd">    Arguments:</span>

<span class="sd">      start : number</span>
<span class="sd">        starting number</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nextnum</span> <span class="o">=</span> <span class="n">start</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">subnum</span> <span class="o">=</span> <span class="mi">0</span>
</div>
<div class="viewcode-block" id="SequenceNumerator.next"><a class="viewcode-back" href="../index.html#profscore.SequenceNumerator.next">[docs]</a>  <span class="k">def</span> <span class="nf">next</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return next (whole) number.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num</span>  <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nextnum</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nextnum</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nextnum</span><span class="o">+</span><span class="mi">1</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">subnum</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">nextnum</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">nextnum</span> <span class="o">=</span> <span class="mi">1</span> <span class="c"># no zeroes, -1 to +1</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">num</span>
</div>
<div class="viewcode-block" id="SequenceNumerator.subnext"><a class="viewcode-back" href="../index.html#profscore.SequenceNumerator.subnext">[docs]</a>  <span class="k">def</span> <span class="nf">subnext</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return next sub number, e.g. 1a, 1b etc.</span>
<span class="sd">    Must be called after a call to ``next()``.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">num</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;SequenceNumerator cannot request a subnumber before a number requested&quot;</span><span class="p">)</span>
    <span class="n">ret</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%i%s</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">num</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">__class__</span><span class="o">.</span><span class="n">subnums</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">subnum</span><span class="p">])</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">subnum</span> <span class="o">+=</span><span class="mi">1</span>
    <span class="k">return</span> <span class="n">ret</span>

</div></div>
<div class="viewcode-block" id="generateGenericReport"><a class="viewcode-back" href="../index.html#profscore.generateGenericReport">[docs]</a><span class="k">def</span> <span class="nf">generateGenericReport</span><span class="p">(</span><span class="n">outputHandle</span><span class="p">,</span> <span class="n">outputObject</span><span class="p">,</span>
  <span class="n">records</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">alignmentPDF</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">referenceId</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span>
  <span class="n">outputPositions</span> <span class="o">=</span> <span class="n">OutputPositions</span><span class="o">.</span><span class="n">PRESENT</span><span class="p">,</span>
  <span class="n">simpleNumbering</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">numberingStart</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
  <span class="n">blockSequenceOutput</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">showCompleteProfile</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">pAsFraction</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a conservation score report</span>

<span class="sd">    Arguments:</span>

<span class="sd">      path : string</span>
<span class="sd">        output path for the report</span>
<span class="sd">      outputObject : ConservationScores or SequenceProfile object</span>
<span class="sd">        scores or profile to output in the report</span>
<span class="sd">      records : list of Bio.SeqRecord objects, optional</span>
<span class="sd">        list of aligned sequences to output, use to override records stored in ``outputObject``</span>
<span class="sd">      alignmentPDF : ScoreProbDist object, optional</span>
<span class="sd">        probability distribution of the scores at each site, used to assign p-values, if absent not p-values written;</span>
<span class="sd">        must be prepared for lookup by precalling ``prepareLookup``</span>
<span class="sd">      referenceId : string, optional</span>
<span class="sd">        sequence record id to use as the reference sequence, used for numbering site positions, if absent first second used for numbering</span>
<span class="sd">      outputPositions : OutputPositions.ALL, OutputPositions.PRESENT or OutputPositions.REFERENCE, optional</span>
<span class="sd">        if ``OutputPositions.ALL`` scores for all positions written,</span>
<span class="sd">        if ``OutputPositions.PRESENT`` (default) only present positions written (i.e. at least one non-gap at site),</span>
<span class="sd">        if ``OutputPositions.REFERENCE`` only positions in reference sequence written (i.e. at least one non-gap at site),</span>
<span class="sd">      simpleNumbering : boolean, optional</span>
<span class="sd">        if ``True`` simple numbering (1,2,3...) used, otherwise (default) numbering is for reference sequence</span>
<span class="sd">      numberingStart : number, optional</span>
<span class="sd">        number used to start numbering, if absent numbering starts from 1</span>
<span class="sd">      blockSequenceOutput : boolean, optional</span>
<span class="sd">        if ``True`` sequences output as a block (column) of text, with one block for all scored sequences and</span>
<span class="sd">        one block for sequences not scored (but added to output), otherwise (default) each sequence is written to individual columns</span>
<span class="sd">      showCompleteProfile : boolean, optional</span>
<span class="sd">        if ``True`` complete sequence profile shown (i.e. score for each amino acid), otherwise (default) profile not shown; only valid for ``ProfileScore`` or ``SequenceProfile``</span>
<span class="sd">      pAsFraction : boolean, optional</span>
<span class="sd">        if ``True`` p-values produced as fraction e.g. 130/1000, otherwise (default) produced as a proportion</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># get record names for header</span>
    <span class="c">#recordNames = numberingBy</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Writing report...&quot;</span>

    <span class="n">reportMetadata</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span> <span class="c"># details about report</span>
    <span class="n">reportMetadata</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">outputObject</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>

    <span class="n">sequenceProfile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">outputObject</span><span class="p">,</span><span class="n">SequenceProfile</span><span class="p">):</span>
      <span class="n">sequenceProfile</span> <span class="o">=</span> <span class="n">outputObject</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">outputObject</span><span class="p">,</span><span class="n">ProfileScore</span><span class="p">):</span>
      <span class="n">sequenceProfile</span> <span class="o">=</span> <span class="n">outputObject</span><span class="o">.</span><span class="n">sequenceProfile</span>
      <span class="n">reportMetadata</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">outputObject</span><span class="o">.</span><span class="n">sequenceProfile</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">records</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span> <span class="n">records</span> <span class="o">=</span> <span class="n">outputObject</span><span class="o">.</span><span class="n">records</span>
    <span class="k">if</span> <span class="n">referenceId</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span> <span class="n">referenceId</span> <span class="o">=</span> <span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>

    <span class="c"># determine index of record to number by</span>
    <span class="n">referenceIndex</span> <span class="o">=</span> <span class="p">[</span><span class="n">index</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">record</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">records</span><span class="p">)</span> <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">referenceId</span> <span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">referenceIndex</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&quot;Couldn&#39;t find reference sequence </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">referenceId</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span> <span class="n">referenceIndex</span> <span class="o">=</span> <span class="n">referenceIndex</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Reference sequence is </span><span class="si">%s</span><span class="s">, index </span><span class="si">%i</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">referenceId</span><span class="p">,</span><span class="n">referenceIndex</span><span class="p">)</span>

    <span class="n">referenceRecord</span> <span class="o">=</span> <span class="n">records</span><span class="p">[</span><span class="n">referenceIndex</span><span class="p">]</span>

    <span class="c"># find output sites</span>
    <span class="k">if</span> <span class="n">outputPositions</span> <span class="o">==</span> <span class="n">OutputPositions</span><span class="o">.</span><span class="n">ALL</span><span class="p">:</span>
        <span class="n">sitesOfInterest</span> <span class="o">=</span> <span class="n">getAllSites</span><span class="p">(</span><span class="n">records</span><span class="p">)</span>
        <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report site selections&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;all sites&#39;</span>
    <span class="k">elif</span> <span class="n">outputPositions</span> <span class="o">==</span> <span class="n">OutputPositions</span><span class="o">.</span><span class="n">REFERENCE</span><span class="p">:</span>
      <span class="n">sitesOfInterest</span> <span class="o">=</span> <span class="n">getPresentSites</span><span class="p">([</span><span class="n">records</span><span class="p">[</span><span class="n">referenceIndex</span><span class="p">]])</span>
      <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report site selections&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;sites in </span><span class="si">%s</span><span class="s">&#39;</span><span class="o">%</span><span class="n">referenceId</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">sitesOfInterest</span> <span class="o">=</span> <span class="n">getPresentSites</span><span class="p">(</span><span class="n">records</span><span class="p">)</span>
      <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report site selections&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;present sites&#39;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Writing report with </span><span class="si">%i</span><span class="s"> sites.&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">))</span>

    <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report sites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;</span><span class="si">%i</span><span class="s"> sites&#39;</span><span class="o">%</span><span class="nb">len</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">)</span>
    <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report sequences shown&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">)</span>


    <span class="c"># get numbering</span>
    <span class="k">if</span> <span class="n">simpleNumbering</span><span class="p">:</span>
      <span class="n">sequenceNumerator</span> <span class="o">=</span> <span class="n">SequenceNumerator</span><span class="p">(</span><span class="n">numberingStart</span><span class="p">)</span>
      <span class="n">sequenceNums</span> <span class="o">=</span> <span class="p">[</span><span class="n">sequenceNumerator</span><span class="o">.</span><span class="n">next</span><span class="p">()</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">sitesOfInterest</span><span class="p">]</span>
      <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report numbering by&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;simple numbering&#39;</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="c"># number by reference sequence</span>

      <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">referenceRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span><span class="o">!=</span><span class="s">&#39;-&#39;</span><span class="p">:</span>
          <span class="k">break</span> <span class="c"># find first non-gap residue in reference</span>

      <span class="n">firstNonGap</span> <span class="o">=</span> <span class="n">site</span> <span class="c"># number of first site</span>
      <span class="n">firstNonGapInSOI</span> <span class="o">=</span> <span class="n">index</span> <span class="c"># index of first site within sitesOfInterest</span>
      <span class="n">firstNonGapNum</span> <span class="o">=</span> <span class="n">numberingStart</span>

      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">3</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;First non-gap site is </span><span class="si">%i</span><span class="s">, numbered </span><span class="si">%i</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">numberingStart</span><span class="p">)</span>

      <span class="c"># determine sequence numbering</span>
      <span class="c"># number: -3, -2, -1, 1, 2, 3, 3a, 3b, 3c, 4</span>
      <span class="c"># a,b,c used if there are gaps in reference sequence</span>

      <span class="n">lastGapNum</span> <span class="o">=</span> <span class="n">numberingStart</span> <span class="o">-</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">numberingStart</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="o">-</span><span class="mi">1</span> <span class="c"># sites up to first non-gap always given negative numbers</span>
      <span class="n">sequenceNumerator</span> <span class="o">=</span> <span class="n">SequenceNumerator</span><span class="p">(</span><span class="n">lastGapNum</span><span class="o">+</span><span class="mi">1</span><span class="o">-</span><span class="n">firstNonGapInSOI</span><span class="p">)</span> <span class="c"># init numerator to first site number, so that sites up to first non gap -x... -2,-1</span>
      <span class="n">sequenceNums</span> <span class="o">=</span> <span class="p">[]</span>
      <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">sitesOfInterest</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">site</span><span class="o">==</span><span class="n">firstNonGap</span><span class="p">:</span> <span class="n">sequenceNumerator</span><span class="o">.</span><span class="n">reset</span><span class="p">(</span><span class="n">firstNonGapNum</span><span class="p">)</span> <span class="c"># jump to first numbering position (only has effect if this is &gt;1)</span>
        <span class="k">if</span> <span class="n">site</span><span class="o">&lt;</span><span class="n">firstNonGap</span> <span class="ow">or</span> <span class="n">referenceRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span><span class="o">!=</span><span class="s">&#39;-&#39;</span><span class="p">:</span>
          <span class="n">nextNum</span> <span class="o">=</span> <span class="n">sequenceNumerator</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
          <span class="n">nextNum</span> <span class="o">=</span> <span class="n">sequenceNumerator</span><span class="o">.</span><span class="n">subnext</span><span class="p">()</span>
        <span class="n">sequenceNums</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nextNum</span><span class="p">)</span>
      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&quot;Site numbering (index, refseq, site num):&quot;</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">siteNum</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequenceNums</span><span class="p">):</span> <span class="k">print</span> <span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">referenceRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">sitesOfInterest</span><span class="p">[</span><span class="n">index</span><span class="p">]],</span><span class="n">siteNum</span>
      <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report numbering by&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;reference sequence </span><span class="si">%s</span><span class="s"> starting from </span><span class="si">%i</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">referenceId</span><span class="p">,</span><span class="n">numberingStart</span><span class="p">)</span>
    <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;report numbering extent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;</span><span class="si">%s</span><span class="s"> to </span><span class="si">%s</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">sequenceNums</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span><span class="nb">str</span><span class="p">(</span><span class="n">sequenceNums</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>

    <span class="n">reportRows</span> <span class="o">=</span> <span class="p">[</span><span class="n">OrderedDict</span><span class="p">()</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">sitesOfInterest</span><span class="p">]</span>

    <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;note sequences&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;sequences marked * were not scored&quot;</span>

    <span class="n">refSeqHeader</span> <span class="o">=</span> <span class="s">&#39;</span><span class="si">%s</span><span class="s">Ref: </span><span class="si">%s</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="s">&#39;&#39;</span> <span class="k">if</span> <span class="n">referenceRecord</span> <span class="ow">in</span> <span class="n">outputObject</span><span class="o">.</span><span class="n">records</span> <span class="k">else</span> <span class="s">&#39;*&#39;</span><span class="p">,</span> <span class="n">referenceId</span><span class="p">)</span>

    <span class="c"># add data for numbering</span>
    <span class="k">for</span> <span class="n">siteIndex</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">):</span>
      <span class="n">reportRow</span> <span class="o">=</span> <span class="n">reportRows</span><span class="p">[</span><span class="n">siteIndex</span><span class="p">]</span>
      <span class="n">reportRow</span><span class="p">[</span><span class="s">&#39;alignment num&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">site</span><span class="o">+</span><span class="mi">1</span> <span class="c">#1-based</span>
      <span class="n">reportRow</span><span class="p">[</span><span class="s">&#39;report num&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequenceNums</span><span class="p">[</span><span class="n">siteIndex</span><span class="p">]</span>

    <span class="c"># add data for reference seq</span>
    <span class="k">for</span> <span class="n">siteIndex</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">):</span>
      <span class="n">reportRow</span> <span class="o">=</span> <span class="n">reportRows</span><span class="p">[</span><span class="n">siteIndex</span><span class="p">]</span>
      <span class="n">reportRow</span><span class="p">[</span><span class="n">refSeqHeader</span><span class="p">]</span> <span class="o">=</span> <span class="n">referenceRecord</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>

    <span class="c"># add data for other sequences</span>
    <span class="k">if</span> <span class="n">blockSequenceOutput</span><span class="p">:</span>
      <span class="c"># write sequences as a block of text</span>
      <span class="n">recordsNotInScoreHeader</span> <span class="o">=</span> <span class="s">&#39;*Not scored:&#39;</span>
      <span class="n">recordsInScoreHeader</span> <span class="o">=</span> <span class="s">&#39;Scored:&#39;</span>

      <span class="n">indexS</span> <span class="o">=</span> <span class="mi">1</span>
      <span class="n">indexNS</span> <span class="o">=</span> <span class="mi">1</span>
      <span class="c"># generate headers</span>
      <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">outputObject</span><span class="o">.</span><span class="n">records</span><span class="p">:</span>
          <span class="c"># write scored sequence names to header</span>
          <span class="n">recordsInScoreHeader</span> <span class="o">+=</span> <span class="s">&#39; </span><span class="si">%i</span><span class="s">=</span><span class="si">%s</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">indexS</span><span class="p">,</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
          <span class="n">indexS</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
          <span class="c"># write not-scored sequence names to header</span>
          <span class="n">recordsNotInScoreHeader</span> <span class="o">+=</span> <span class="s">&#39; </span><span class="si">%i</span><span class="s">=</span><span class="si">%s</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">indexNS</span><span class="p">,</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
          <span class="n">indexNS</span> <span class="o">+=</span> <span class="mi">1</span>

      <span class="c"># write sequences</span>
      <span class="k">for</span> <span class="n">siteIndex</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">):</span>
        <span class="n">reportRow</span> <span class="o">=</span> <span class="n">reportRows</span><span class="p">[</span><span class="n">siteIndex</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">outputObject</span><span class="o">.</span><span class="n">records</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">reportRow</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">recordsInScoreHeader</span><span class="p">,</span><span class="s">&#39;&#39;</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">+=</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
            <span class="n">reportRow</span><span class="p">[</span><span class="n">recordsInScoreHeader</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

          <span class="k">else</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">reportRow</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">recordsNotInScoreHeader</span><span class="p">,</span><span class="s">&#39;&#39;</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">+=</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
            <span class="n">reportRow</span><span class="p">[</span><span class="n">recordsNotInScoreHeader</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">blockSequenceOutput</span><span class="p">:</span> <span class="c"># write sequences as individual columns</span>
      <span class="c"># write sequences</span>
      <span class="k">for</span> <span class="n">siteIndex</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">):</span>
        <span class="n">reportRow</span> <span class="o">=</span> <span class="n">reportRows</span><span class="p">[</span><span class="n">siteIndex</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
          <span class="c"># write not scored first</span>
          <span class="k">if</span> <span class="n">record</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">outputObject</span><span class="o">.</span><span class="n">records</span><span class="p">:</span> <span class="n">keyname</span> <span class="o">=</span> <span class="s">&#39;*&#39;</span><span class="o">+</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="c"># add * to indicate not scored</span>
          <span class="k">else</span><span class="p">:</span> <span class="n">keyname</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span>
          <span class="k">while</span> <span class="n">keyname</span> <span class="ow">in</span> <span class="n">reportRow</span><span class="p">:</span> <span class="n">keyname</span><span class="o">+=</span><span class="s">&quot; &quot;</span> <span class="c"># deal with duplicates by adding spaces to name</span>

          <span class="n">reportRow</span><span class="p">[</span><span class="n">keyname</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>

    <span class="c"># output sequence profile</span>
    <span class="k">if</span> <span class="n">sequenceProfile</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">showCompleteProfile</span><span class="p">:</span>
        <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;note profile&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;showing complete profile vector, score for each amino acid listed according to one letter code&quot;</span>

      <span class="k">if</span> <span class="n">sequenceProfile</span><span class="o">.</span><span class="n">consensus</span><span class="p">:</span>
          <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;note consensus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;showing consensus determined from profile&quot;</span>
          <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;consensus details&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequenceProfile</span><span class="o">.</span><span class="n">consensusDetails</span>

      <span class="n">letters</span> <span class="o">=</span> <span class="n">sequenceProfile</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">letters</span>
      <span class="k">for</span> <span class="n">siteIndex</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">):</span>
        <span class="n">reportRow</span> <span class="o">=</span> <span class="n">reportRows</span><span class="p">[</span><span class="n">siteIndex</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">showCompleteProfile</span><span class="p">:</span>
          <span class="n">scoresAtSite</span> <span class="o">=</span> <span class="n">sequenceProfile</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
          <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="n">letters</span><span class="p">:</span>
              <span class="n">letterIndex</span> <span class="o">=</span> <span class="n">sequenceProfile</span><span class="o">.</span><span class="n">vectorSubsMat</span><span class="o">.</span><span class="n">alphabetLUT</span><span class="p">[</span><span class="nb">ord</span><span class="p">(</span><span class="n">letter</span><span class="p">)]</span>
              <span class="n">reportRow</span><span class="p">[</span><span class="n">letter</span><span class="p">]</span> <span class="o">=</span> <span class="nb">repr</span><span class="p">(</span><span class="n">scoresAtSite</span><span class="p">[</span><span class="n">letterIndex</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">sequenceProfile</span><span class="o">.</span><span class="n">consensus</span><span class="p">:</span>
            <span class="n">consensusAtSite</span> <span class="o">=</span> <span class="n">sequenceProfile</span><span class="o">.</span><span class="n">consensus</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">consensusAtSite</span><span class="p">):</span>
              <span class="n">consensusHeader</span> <span class="o">=</span> <span class="s">&#39;consensus </span><span class="si">%i</span><span class="s">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
              <span class="n">reportRow</span><span class="p">[</span><span class="n">consensusHeader</span><span class="p">]</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">letter</span>
              <span class="n">reportRow</span><span class="p">[</span><span class="n">consensusHeader</span><span class="o">+</span><span class="s">&quot; score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">repr</span><span class="p">(</span><span class="n">aa</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>

    <span class="c"># output scores</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">outputObject</span><span class="p">,</span><span class="n">ConservationScores</span><span class="p">):</span>
      <span class="n">conservationScores</span> <span class="o">=</span> <span class="n">outputObject</span>
      <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;note scores&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;showing score for each position under score&quot;</span>
      <span class="k">if</span> <span class="n">alignmentPDF</span><span class="p">:</span>
        <span class="c">#reportMetadata[&#39;note p-value&#39;] = &quot;showing p-value under pvalue, determined from %i randomizations&quot;%alignmentPDF.totalRandomizations</span>
        <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;note p-value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;showing p-value from randomizations under pvalue&quot;</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">prepareMetadata</span><span class="p">()</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span> <span class="c"># add randomization metadata</span>
          <span class="k">if</span> <span class="n">key</span><span class="o">!=</span><span class="n">AlignmentPDF</span><span class="o">.</span><span class="n">dataMagicWord</span><span class="p">:</span>
            <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;randomization &#39;</span><span class="o">+</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
      <span class="k">for</span> <span class="n">siteIndex</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sitesOfInterest</span><span class="p">):</span>
        <span class="n">reportRow</span> <span class="o">=</span> <span class="n">reportRows</span><span class="p">[</span><span class="n">siteIndex</span><span class="p">]</span>
        <span class="n">score</span> <span class="o">=</span> <span class="n">conservationScores</span><span class="o">.</span><span class="n">scores</span> <span class="p">[</span><span class="n">site</span><span class="p">]</span>
        <span class="n">reportRow</span><span class="p">[</span><span class="s">&quot;score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">repr</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">alignmentPDF</span><span class="p">:</span>
          <span class="n">density</span><span class="p">,</span><span class="n">totalDensity</span> <span class="o">=</span> <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">lookupDensities</span><span class="p">(</span><span class="n">site</span><span class="p">,</span><span class="n">score</span><span class="p">)</span>
          <span class="k">if</span> <span class="n">pAsFraction</span><span class="p">:</span> <span class="n">reportRow</span><span class="p">[</span><span class="s">&quot;pvalue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%i</span><span class="s">/</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">density</span><span class="p">,</span><span class="n">totalDensity</span><span class="p">)</span>
          <span class="k">else</span><span class="p">:</span> <span class="n">reportRow</span><span class="p">[</span><span class="s">&quot;pvalue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">density</span><span class="o">*</span><span class="mf">1.0</span><span class="o">/</span><span class="n">totalDensity</span><span class="p">)</span>

    <span class="c"># write metadata</span>
    <span class="n">csvWriter</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">outputHandle</span><span class="p">)</span>
    <span class="n">reportMetadata</span><span class="p">[</span><span class="s">&#39;note header ends&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;data follows immediately below&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Writing metadata...&#39;</span>
    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">reportMetadata</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;: &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">item</span><span class="p">])</span>
      <span class="n">csvWriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>

    <span class="c"># write rows</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">1</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Writing data...&#39;</span>

    <span class="n">referenceRow</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">for</span> <span class="n">reportRow</span> <span class="ow">in</span> <span class="n">reportRows</span><span class="p">:</span> <span class="c"># use largest row as reference row</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">referenceRow</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">reportRow</span><span class="p">)</span><span class="o">&gt;</span><span class="nb">len</span><span class="p">(</span><span class="n">referenceRow</span><span class="p">):</span>
        <span class="n">referenceRow</span> <span class="o">=</span> <span class="n">reportRow</span>


    <span class="n">csvDictWrite</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictWriter</span><span class="p">(</span><span class="n">outputHandle</span><span class="p">,</span> <span class="n">referenceRow</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Reference row (headings) has </span><span class="si">%i</span><span class="s"> columns.&#39;</span><span class="o">%</span><span class="nb">len</span><span class="p">(</span><span class="n">referenceRow</span><span class="p">)</span>
    <span class="n">csvDictWrite</span><span class="o">.</span><span class="n">writeheader</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">reportRow</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reportRows</span><span class="p">):</span>
      <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">4</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Row </span><span class="si">%i</span><span class="s"> has </span><span class="si">%i</span><span class="s"> columns.&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">reportRow</span><span class="p">))</span>
      <span class="n">csvDictWrite</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">reportRow</span><span class="p">)</span>

    <span class="n">outputHandle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span>



<span class="c">#@profile</span></div>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../index.html#profscore.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
  <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Execute main script code function.</span>

<span class="sd">  Verbosity levels:</span>

<span class="sd">    0/None : default, main messages</span>

<span class="sd">    1 : a few extra messages showing progress</span>

<span class="sd">    2 : outputs top level calculations</span>

<span class="sd">    3 : outputs mid-level calculations</span>

<span class="sd">    4 : outputs detailed calculations</span>
<span class="sd">  &quot;&quot;&quot;</span>

  <span class="k">global</span> <span class="n">verbose</span>
  <span class="k">global</span> <span class="n">version</span>

  <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">formatter_class</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span><span class="p">,</span>
                                   <span class="n">description</span><span class="o">=</span><span class="s">&#39;Calculates sequence conservation scores&#39;</span><span class="p">,</span>
                                   <span class="n">epilog</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
  <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;infile&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;input alignment to process, fasta format&quot;</span><span class="p">)</span>

  <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-v&quot;</span><span class="p">,</span> <span class="s">&quot;--verbose&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&quot;count&quot;</span><span class="p">,</span><span class="n">help</span><span class="o">=</span><span class="s">&quot;produce verbose output, increase number of times specified to increase verbosity&quot;</span><span class="p">)</span>
  <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;--version&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&#39;version&#39;</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="s">&#39;</span><span class="si">%(prog)s</span><span class="s"> &#39;</span><span class="o">+</span><span class="s">&quot;</span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">version</span><span class="p">)</span>
  <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;--overwrite&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&#39;store_true&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;force overwriting of output files&quot;</span><span class="p">)</span>

  <span class="n">scoreGroup</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&#39;score parameters&#39;</span><span class="p">,</span> <span class="s">&#39;parameters for score calculation&#39;</span><span class="p">)</span>

  <span class="n">scoreGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-s&quot;</span><span class="p">,</span> <span class="s">&quot;--subset&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;.*&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;subset of sequences to analyse, selected by regular expression search for SUBSET&quot;</span><span class="p">)</span>
  <span class="n">scoreGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-t&quot;</span><span class="p">,</span> <span class="s">&quot;--type&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;PROFILE&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;type of conservation score calculated, VALDAR for Valdar score, PROFILE for (corrected) profile-based score, PROFILE_UNCORR for uncorrected profile-based score&quot;</span><span class="p">,</span>
    <span class="n">choices</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;VALDAR&quot;</span><span class="p">,</span><span class="s">&quot;PROFILE&quot;</span><span class="p">,</span><span class="s">&quot;PROFILE_UNCORR&quot;</span><span class="p">))</span>
  <span class="n">scoreGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-m&quot;</span><span class="p">,</span><span class="s">&quot;--subsmat&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;Valdar2001&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;substitution matrix to use, aaindex2 format&quot;</span><span class="p">)</span>
  <span class="n">scoreGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--diagonal&quot;</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;score to use for diagonal elements in substitution matrix (identities), set prior to matrix transformation, if none the rounded average of diagonal elements in the unmodified matrix used&quot;</span><span class="p">)</span>
  <span class="n">scoreGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--gap&quot;</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span><span class="n">default</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;score to use for gaps, set after matrix transformation&quot;</span><span class="p">)</span>
  <span class="n">scoreGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--gapgap&quot;</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;score to use for gap-gap substitutions, set after matrix transformation, if absent GAP score used&quot;</span><span class="p">)</span>
  <span class="n">scoreGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--uniform&quot;</span><span class="p">,</span><span class="n">action</span><span class="o">=</span><span class="s">&#39;store_true&#39;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;if present, apply weighting to all sequences, otherwise sequences weighted by evolutionary distance&quot;</span><span class="p">)</span>

  <span class="n">consGroup</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&#39;consensus parameters&#39;</span><span class="p">,</span> <span class="s">&#39;parameters for determining consensus residues (profile-based scores only)&#39;</span><span class="p">)</span>

  <span class="n">consGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-c&quot;</span><span class="p">,</span><span class="s">&quot;--cutoff&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;score cutoff for consensus residues, if absent consensus not determined&quot;</span><span class="p">)</span>
  <span class="n">consGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--propcutoff&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span><span class="n">default</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;proportional score cutoff for consensus residues consensus may also be this proportion of the highest scoring residue&quot;</span><span class="p">)</span>

  <span class="n">rndGroup</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&#39;randomization parameters&#39;</span><span class="p">,</span> <span class="s">&#39;parameters for randomizations&#39;</span><span class="p">)</span>

  <span class="n">rndGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-n&quot;</span><span class="p">,</span><span class="s">&quot;--numrnd&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span><span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;number of (additional) randomizations, if none no randomizations performed, must be none if combining multiple input files&quot;</span><span class="p">)</span>
  <span class="n">rndGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--rndsubset&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;.*&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;subset of sequences to randomize from, selected by regular expression search for RNDSUBSET, default is all sequences&quot;</span><span class="p">)</span>
  <span class="n">rndGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-i&quot;</span><span class="p">,</span><span class="s">&quot;--rndinfile&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s">&#39;*&#39;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;randomization input file(s) for reading previously performed randomizations, if multiple files specified these will be combined into output file&quot;</span><span class="p">)</span>
  <span class="n">rndGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-p&quot;</span><span class="p">,</span><span class="s">&quot;--rndoutfile&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;randomization output file, required when performing additional randomizations or combining multiple input files, must differ from RNDINFILE&quot;</span><span class="p">)</span>

  <span class="n">outputGroup</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&#39;output parameters&#39;</span><span class="p">,</span> <span class="s">&#39;parameters for generating score report&#39;</span><span class="p">)</span>


  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-o&quot;</span><span class="p">,</span><span class="s">&quot;--outfile&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;output file to produce, csv format, if not present no report generated&quot;</span><span class="p">)</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-r&quot;</span><span class="p">,</span><span class="s">&quot;--refid&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;reference sequence for numbering, if absent first sequence in subset used&quot;</span><span class="p">)</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--outsubset&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;subset of sequences to output in report, selected by regular expression search for OUTSUBSET, if absent score SUBSET used&quot;</span><span class="p">)</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--outpos&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;PRESENT&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;positions to output in report, PRESENT for non-gap postions, ALL for all positions or REFERENCE for positions in reference sequence only&quot;</span><span class="p">,</span>
    <span class="n">choices</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;PRESENT&quot;</span><span class="p">,</span><span class="s">&quot;ALL&quot;</span><span class="p">,</span><span class="s">&quot;REFERENCE&quot;</span><span class="p">))</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--simplenum&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&#39;store_true&#39;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;if present use simple sequential numbering of sequence positions (1,2,3...), otherwise numbering is by reference sequence&quot;</span><span class="p">)</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--numstart&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;start sequence numbering at this value&quot;</span><span class="p">)</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--blockoutput&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&#39;store_true&#39;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;if present output sequences in as a single block of text, otherwise default to one column per sequence&quot;</span><span class="p">)</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--completeprofile&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&#39;store_true&#39;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;if present output complete sequence profile, i.e. profile scores for all positions&quot;</span><span class="p">)</span>
  <span class="n">outputGroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--pfraction&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&#39;store_true&#39;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s">&quot;if present output p-values as fractions instead of proportions&quot;</span><span class="p">)</span>


  <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

  <span class="k">print</span> <span class="s">&quot;&quot;</span>
  <span class="k">print</span> <span class="s">&quot;Parameters:&quot;</span>

  <span class="n">verbose</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">verbose</span>
  <span class="n">overwrite</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span>
  <span class="k">print</span> <span class="s">&quot; Input alignment file is </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">infile</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot; Verbosity level </span><span class="si">%i</span><span class="s"> and will </span><span class="si">%s</span><span class="s">overwrite files.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">verbose</span><span class="p">,</span><span class="s">&quot;&quot;</span> <span class="k">if</span> <span class="n">overwrite</span> <span class="k">else</span> <span class="s">&quot;NOT &quot;</span><span class="p">)</span>

  <span class="c"># get score arguments</span>
  <span class="n">scoreOptions</span> <span class="o">=</span> <span class="n">Dummy</span><span class="p">()</span>
  <span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsetRE</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">subset</span>
  <span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">ScoreType</span><span class="p">,</span><span class="n">args</span><span class="o">.</span><span class="n">type</span><span class="p">)</span>
  <span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsMat</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">subsmat</span>
  <span class="n">scoreOptions</span><span class="o">.</span><span class="n">gap</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gap</span><span class="p">)</span>
  <span class="n">scoreOptions</span><span class="o">.</span><span class="n">gapgap</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gapgap</span><span class="p">)</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gapgap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">gap</span>
  <span class="n">scoreOptions</span><span class="o">.</span><span class="n">diag</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">diagonal</span>
  <span class="k">if</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">diag</span><span class="p">:</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">diag</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">diag</span><span class="p">)</span> <span class="c"># convert to float if not none</span>
  <span class="n">scoreOptions</span><span class="o">.</span><span class="n">uniform</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">uniform</span>

  <span class="k">print</span> <span class="s">&quot; Calculating scores from subset </span><span class="se">\&quot;</span><span class="si">%s</span><span class="se">\&quot;</span><span class="s"> using </span><span class="si">%s</span><span class="s"> score.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="p">,</span><span class="n">ScoreType</span><span class="o">.</span><span class="n">reverse_mapping</span><span class="p">[</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span><span class="p">])</span>
  <span class="k">print</span> <span class="s">&quot; Using subsitution matrix </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsMat</span><span class="p">)</span>
  <span class="k">print</span> <span class="s">&quot; Score for gaps is </span><span class="si">%.2f</span><span class="s">, gap-gap is </span><span class="si">%.2f</span><span class="s"> and diagonal is </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">gap</span><span class="p">,</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">gapgap</span><span class="p">,</span><span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span><span class="o">%</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">diag</span> <span class="k">if</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">diag</span> <span class="k">else</span> <span class="s">&quot;average&quot;</span><span class="p">)</span>
  <span class="k">print</span> <span class="s">&quot; </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;Uniform weighting of all sequences&quot;</span> <span class="k">if</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">uniform</span> <span class="k">else</span> <span class="s">&quot;Weighting by evolutionary distance&quot;</span><span class="p">)</span>

  <span class="c"># get consensus arguments</span>
  <span class="n">consOptions</span> <span class="o">=</span> <span class="bp">None</span>
  <span class="k">if</span> <span class="n">isProfileScore</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">cutoff</span><span class="p">:</span>
      <span class="n">consOptions</span> <span class="o">=</span> <span class="n">Dummy</span><span class="p">()</span>
      <span class="n">consOptions</span><span class="o">.</span><span class="n">cutoffAbs</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">cutoff</span><span class="p">)</span>
      <span class="n">consOptions</span><span class="o">.</span><span class="n">cutoffProp</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">propcutoff</span><span class="p">)</span>

  <span class="k">if</span> <span class="n">consOptions</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot; Consensus residues must have score of &gt;=</span><span class="si">%.2f</span><span class="s"> or &gt;=</span><span class="si">%.2f</span><span class="s"> of highest score.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">consOptions</span><span class="o">.</span><span class="n">cutoffAbs</span><span class="p">,</span><span class="n">consOptions</span><span class="o">.</span><span class="n">cutoffProp</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot; Not determining consensus residues.&quot;</span>

  <span class="c"># get randomization arguments</span>
  <span class="n">rndOptions</span> <span class="o">=</span> <span class="n">Dummy</span><span class="p">()</span>
  <span class="n">rndOptions</span><span class="o">.</span><span class="n">number</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">numrnd</span>
  <span class="n">rndOptions</span><span class="o">.</span><span class="n">infile</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">rndinfile</span>
  <span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">rndoutfile</span>
  <span class="n">rndOptions</span><span class="o">.</span><span class="n">subsetRE</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">rndsubset</span>

  <span class="k">print</span> <span class="s">&quot; </span><span class="si">%i</span><span class="s"> randomizations requested from subset </span><span class="se">\&quot;</span><span class="si">%s</span><span class="se">\&quot;</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">number</span><span class="p">,</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">infile</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot; Using randomization input files </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">infile</span><span class="p">))</span>
  <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot; Using randomization output file </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span>

  <span class="c"># get report arguments</span>
  <span class="n">reportOptions</span> <span class="o">=</span> <span class="n">Dummy</span><span class="p">()</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">outfile</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">refSeqID</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">refid</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">subsetRE</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">outsubset</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">outsubset</span> <span class="k">else</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsetRE</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">outputPositions</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">OutputPositions</span><span class="p">,</span><span class="n">args</span><span class="o">.</span><span class="n">outpos</span><span class="p">)</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">simpleNumbering</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">simplenum</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">numStart</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">numstart</span><span class="p">)</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">blockOutput</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">blockoutput</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">showCompleteProfile</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">completeprofile</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">pAsFraction</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">pfraction</span>
  <span class="n">reportOptions</span><span class="o">.</span><span class="n">rndlookupfile</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">rndoutfile</span>
  <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">rndlookupfile</span> <span class="ow">is</span> <span class="bp">None</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">rndinfile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
   <span class="n">reportOptions</span><span class="o">.</span><span class="n">rndlookupfile</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">rndinfile</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
  <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot; Using randomization lookup file </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span>
  <span class="k">if</span> <span class="ow">not</span> <span class="n">isProfileScore</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span><span class="p">):</span>
    <span class="n">reportOptions</span><span class="o">.</span><span class="n">showCompleteProfile</span> <span class="o">=</span> <span class="bp">None</span> <span class="c"># only profile scores can use this option</span>
  <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot; Writing report to </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot; Will write scores from subset </span><span class="se">\&quot;</span><span class="si">%s</span><span class="se">\&quot;</span><span class="s"> as </span><span class="si">%s</span><span class="s"> with </span><span class="si">%s</span><span class="s"> as reference sequence.&quot;</span><span class="o">%</span>\
      <span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="p">,</span> <span class="s">&quot;a block&quot;</span> <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">blockOutput</span> <span class="k">else</span> <span class="s">&quot;columns&quot;</span><span class="p">,</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">refSeqID</span> <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">refSeqID</span> <span class="k">else</span> <span class="s">&quot;first sequence&quot;</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot; Numbering from </span><span class="si">%i</span><span class="s"> using </span><span class="si">%s</span><span class="s"> numbering.&quot;</span><span class="o">%</span>\
      <span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">numStart</span><span class="p">,</span><span class="s">&quot;simple&quot;</span> <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">simpleNumbering</span> <span class="k">else</span> <span class="s">&quot;reference sequence&quot;</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot; Will write </span><span class="si">%s</span><span class="s"> positions</span><span class="si">%s%s</span><span class="s">.&quot;</span><span class="o">%</span>\
      <span class="p">(</span><span class="n">OutputPositions</span><span class="o">.</span><span class="n">reverse_mapping</span><span class="p">[</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">outputPositions</span><span class="p">],</span> <span class="s">&quot;, consensus residues&quot;</span> <span class="k">if</span> <span class="n">consOptions</span> <span class="k">else</span> <span class="s">&quot;&quot;</span><span class="p">,</span>
      <span class="s">&quot;, complete profile&quot;</span> <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">showCompleteProfile</span> <span class="k">else</span> <span class="s">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">rndlookupfile</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot; p-values will be looked up from randomization file </span><span class="si">%s</span><span class="s"> and shown as </span><span class="si">%s</span><span class="s">.&quot;</span>\
      <span class="o">%</span><span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">rndlookupfile</span><span class="p">,</span><span class="s">&quot;fractions&quot;</span> <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">pAsFraction</span> <span class="k">else</span> <span class="s">&quot;proportions&quot;</span><span class="p">)</span>

  <span class="c">#sys.exit()</span>

  <span class="k">print</span> <span class="s">&quot;&quot;</span>
  <span class="k">print</span> <span class="s">&quot;Reading/opening files:&quot;</span>

  <span class="c"># read input alignment</span>
  <span class="n">ALLrecords</span> <span class="o">=</span> <span class="p">[]</span> <span class="c"># stores alignment records</span>

  <span class="k">try</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Reading input alignment </span><span class="si">%s</span><span class="s">...&quot;</span><span class="o">%</span><span class="n">args</span><span class="o">.</span><span class="n">infile</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">infile</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;Input alignment file does not exist.&quot;</span>
        <span class="k">print</span> <span class="n">msg</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="n">infile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">infile</span><span class="p">,</span> <span class="s">&quot;rU&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">record</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">)):</span>
        <span class="n">ALLrecords</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;Read sequence </span><span class="si">%i</span><span class="s">: </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">print</span> <span class="s">&quot;Read input alignment containing </span><span class="si">%i</span><span class="s"> sequences.&quot;</span><span class="o">%</span><span class="nb">len</span><span class="p">(</span><span class="n">ALLrecords</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&quot;List of input sequences:&quot;</span>
      <span class="n">printRecordNames</span><span class="p">(</span><span class="n">ALLrecords</span><span class="p">)</span>

  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Error reading input file!&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
      <span class="k">print</span> <span class="n">traceback</span><span class="o">.</span><span class="n">format_exc</span><span class="p">()</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>

  <span class="c"># ensure output file is writable</span>
  <span class="n">outfile</span> <span class="o">=</span> <span class="bp">None</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&quot;Opening output file </span><span class="si">%s</span><span class="s">...&quot;</span><span class="o">%</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">overwrite</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span><span class="o">.</span><span class="n">st_size</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
          <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;Output file exists and contains data. Will not replace.&quot;</span>
          <span class="k">print</span> <span class="n">msg</span>
          <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
      <span class="n">outfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">,</span> <span class="s">&quot;wb&quot;</span><span class="p">)</span>
      <span class="k">print</span> <span class="s">&quot;Output file opened.&quot;</span>
  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Error opening output file!&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
      <span class="k">print</span> <span class="n">traceback</span><span class="o">.</span><span class="n">format_exc</span><span class="p">()</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>

  <span class="c"># read and transform subs matrix</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="c">#print &quot;Reading substitution matrx %s...&quot;%scoreOptions.subsMat</span>
    <span class="n">subsMat</span> <span class="o">=</span> <span class="n">readSubsMat</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsMat</span><span class="p">)</span>
  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Error reading substitution matrix!&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
      <span class="k">print</span> <span class="n">traceback</span><span class="o">.</span><span class="n">format_exc</span><span class="p">()</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>


  <span class="c"># read input randomization file</span>
  <span class="n">rndinfile</span> <span class="o">=</span> <span class="bp">None</span>
  <span class="n">rndinfiles</span> <span class="o">=</span> <span class="bp">None</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">infile</span><span class="p">:</span>
      <span class="n">rndinfiles</span> <span class="o">=</span> <span class="p">[]</span>
      <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">infile</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&quot;Opening randomization input file </span><span class="si">%s</span><span class="s">...&quot;</span><span class="o">%</span><span class="n">path</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;Randomization input file does not exist.&quot;</span>
            <span class="k">print</span> <span class="n">msg</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="n">rndinfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s">&quot;rb&quot;</span><span class="p">)</span>
        <span class="n">rndinfiles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rndinfile</span><span class="p">)</span>
      <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">rndinfiles</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">rndinfiles</span> <span class="o">=</span> <span class="bp">None</span> <span class="c"># use rndinfile</span>
      <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">rndinfiles</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">rndinfile</span> <span class="o">=</span> <span class="bp">None</span> <span class="c"># use rndinfiles</span>
        <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">number</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
          <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;Randomizations requested and multiple input files provided: unsupported.&quot;</span>
          <span class="k">print</span> <span class="n">msg</span>
          <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Error opening randomization input file(s)!&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
      <span class="k">print</span> <span class="n">traceback</span><span class="o">.</span><span class="n">format_exc</span><span class="p">()</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>

  <span class="c"># ensure output randomization file is writable</span>
  <span class="n">rndoutfile</span> <span class="o">=</span> <span class="bp">None</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">number</span><span class="o">&gt;</span><span class="mi">0</span> <span class="ow">or</span> <span class="n">rndinfiles</span><span class="p">:</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;Randomizations or combining input files requested and no randomization output file provided.&quot;</span>
        <span class="k">print</span> <span class="n">msg</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
      <span class="k">print</span> <span class="s">&quot;Opening randomization output file </span><span class="si">%s</span><span class="s">...&quot;</span><span class="o">%</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">overwrite</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span><span class="o">.</span><span class="n">st_size</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
          <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;Randomization output file exists and contains data. Will not replace.&quot;</span>
          <span class="k">print</span> <span class="n">msg</span>
          <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
      <span class="n">rndoutfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">outfile</span><span class="p">,</span> <span class="s">&quot;wb&quot;</span><span class="p">)</span>
  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Error opening randomization output file!&quot;</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
      <span class="k">print</span> <span class="n">traceback</span><span class="o">.</span><span class="n">format_exc</span><span class="p">()</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>


  <span class="c"># Calculate scores for sequence subset</span>
  <span class="k">print</span> <span class="s">&quot;&quot;</span>
  <span class="k">print</span> <span class="s">&quot;Processing:&quot;</span>

  <span class="k">print</span> <span class="s">&quot;Subsetting sequences by score regex </span><span class="si">%s</span><span class="s">...&quot;</span><span class="o">%</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsetRE</span>
  <span class="n">scoreSubsetRecords</span> <span class="o">=</span> <span class="n">subsetByID</span><span class="p">(</span><span class="n">ALLrecords</span><span class="p">,</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="p">)</span>

  <span class="k">print</span> <span class="s">&quot;Got a subset of </span><span class="si">%i</span><span class="s"> sequences.&quot;</span><span class="o">%</span><span class="nb">len</span><span class="p">(</span><span class="n">scoreSubsetRecords</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;List of sequences in subset:&quot;</span>
    <span class="n">printRecordNames</span><span class="p">(</span><span class="n">scoreSubsetRecords</span><span class="p">)</span>

  <span class="k">print</span> <span class="s">&quot;Transforming substitution matrix...&quot;</span>
  <span class="n">linearTransform</span><span class="p">(</span><span class="n">subsMat</span><span class="p">,</span><span class="n">gapgap</span><span class="o">=</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">gapgap</span><span class="p">,</span><span class="n">gap</span><span class="o">=</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">gap</span><span class="p">,</span><span class="n">diagonal</span><span class="o">=</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">diag</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">isProfileScore</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span><span class="p">):</span>
    <span class="k">print</span> <span class="s">&quot;Converting substitution matrix to vector form...&quot;</span>
    <span class="n">vectorSubsMat</span> <span class="o">=</span> <span class="n">VectorSubsMat</span><span class="p">(</span><span class="n">subsMat</span><span class="p">)</span>

  <span class="k">if</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">uniform</span><span class="p">:</span>
    <span class="c"># use simple uniform distances</span>
    <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">createUniformDistanceMatrix</span><span class="p">(</span><span class="n">ALLrecords</span><span class="p">,</span><span class="mf">1.0</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="c"># find full distance matrix, i.e. pairwise distance of all sequences</span>
    <span class="k">print</span> <span class="s">&quot;Calculating pairwise evolutionary distances...&quot;</span>
    <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">calculateDistanceMatrix</span><span class="p">(</span><span class="n">ALLrecords</span><span class="p">,</span><span class="n">subsMat</span><span class="p">)</span>

  <span class="k">print</span> <span class="s">&quot;Calculating scores for subset...&quot;</span>

  <span class="c"># find weights for each sequence from distance matrix</span>
  <span class="k">print</span> <span class="s">&quot;Determining sequence weights for subset...&quot;</span>
  <span class="n">weights</span> <span class="o">=</span> <span class="n">calculateWeights</span><span class="p">(</span><span class="n">scoreSubsetRecords</span><span class="p">,</span> <span class="n">distanceMatrix</span><span class="p">)</span>

  <span class="k">if</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">VALDAR</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Determining Valdar conservation scores for subset...&quot;</span>
    <span class="n">theScores</span> <span class="o">=</span> <span class="n">ValdarScore</span><span class="p">(</span><span class="n">scoreSubsetRecords</span><span class="p">,</span> <span class="n">subsMat</span><span class="p">,</span> <span class="n">weights</span> <span class="o">=</span> <span class="n">weights</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE</span> <span class="ow">or</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE_UNCORR</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&quot;Determining sequence profile for subset...&quot;</span>
      <span class="n">theSequenceProfile</span> <span class="o">=</span> <span class="n">SequenceProfile</span><span class="p">(</span><span class="n">scoreSubsetRecords</span><span class="p">,</span> <span class="n">vectorSubsMat</span><span class="p">,</span> <span class="n">profileWeights</span> <span class="o">=</span> <span class="n">weights</span><span class="p">)</span>
      <span class="k">if</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span> <span class="o">==</span> <span class="n">ScoreType</span><span class="o">.</span><span class="n">PROFILE</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&quot;Determining corrected profile-based conservation scores for subset...&quot;</span>
        <span class="n">correctForSelfComparisons</span> <span class="o">=</span> <span class="bp">True</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&quot;Determining uncorrected profile-based conservation scores for subset...&quot;</span>
        <span class="n">correctForSelfComparisons</span> <span class="o">=</span> <span class="bp">False</span>
      <span class="n">theScores</span> <span class="o">=</span> <span class="n">ProfileScore</span><span class="p">(</span><span class="n">scoreSubsetRecords</span><span class="p">,</span> <span class="n">theSequenceProfile</span> <span class="p">,</span> <span class="n">scoringWeights</span> <span class="o">=</span> <span class="n">weights</span><span class="p">,</span> <span class="n">correctForSelfComparisons</span><span class="o">=</span><span class="n">correctForSelfComparisons</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">AssertionError</span><span class="p">(</span><span class="s">&quot;Invalid score type.&quot;</span><span class="p">)</span>

  <span class="k">if</span> <span class="n">verbose</span><span class="o">&gt;=</span><span class="mi">3</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">site</span><span class="p">,</span><span class="n">score</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">theScores</span><span class="o">.</span><span class="n">scores</span><span class="p">):</span>
      <span class="k">print</span> <span class="s">&quot;Site </span><span class="si">%i</span><span class="s">: score </span><span class="si">%.4f</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">site</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">score</span><span class="p">)</span>

  <span class="k">if</span> <span class="n">consOptions</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;Determining consensus residues...&quot;</span>
    <span class="n">theSequenceProfile</span><span class="o">.</span><span class="n">findConsensus</span><span class="p">(</span><span class="n">consOptions</span><span class="o">.</span><span class="n">cutoffAbs</span><span class="p">,</span> <span class="n">consOptions</span><span class="o">.</span><span class="n">cutoffProp</span><span class="p">)</span>

  <span class="c"># Do randomizations</span>
  <span class="n">alignmentPDF</span> <span class="o">=</span> <span class="bp">None</span>
  <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">number</span><span class="o">&gt;</span><span class="mi">0</span> <span class="ow">or</span> <span class="n">rndinfile</span> <span class="ow">or</span> <span class="n">rndinfiles</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;&quot;</span>
    <span class="k">print</span> <span class="s">&quot;Determining score PDFs:&quot;</span>
    <span class="n">sampleSize</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">scoreSubsetRecords</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot;Subsetting sequences by randomization regex </span><span class="si">%s</span><span class="s">...&quot;</span><span class="o">%</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">subsetRE</span>
    <span class="n">rndSubsetRecords</span> <span class="o">=</span> <span class="n">subsetByID</span><span class="p">(</span><span class="n">ALLrecords</span><span class="p">,</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&quot;List of sequences in randomization subset:&quot;</span>
      <span class="n">printRecordNames</span><span class="p">(</span><span class="n">rndSubsetRecords</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot;Alignment PDFs from randomly sampling </span><span class="si">%i</span><span class="s"> sequences from </span><span class="si">%i</span><span class="s"> sequences...&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">sampleSize</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">rndSubsetRecords</span><span class="p">))</span>
    <span class="n">rndSubsMat</span> <span class="o">=</span> <span class="n">vectorSubsMat</span> <span class="k">if</span> <span class="n">isProfileScore</span><span class="p">(</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span><span class="p">)</span> <span class="k">else</span> <span class="n">subsMat</span>
    <span class="n">alignmentPDF</span> <span class="o">=</span> <span class="n">AlignmentPDF</span><span class="p">(</span><span class="n">records</span> <span class="o">=</span> <span class="n">rndSubsetRecords</span><span class="p">,</span> <span class="n">sampleSize</span> <span class="o">=</span> <span class="n">sampleSize</span><span class="p">,</span> <span class="n">scoreType</span> <span class="o">=</span> <span class="n">scoreOptions</span><span class="o">.</span><span class="n">scoreType</span><span class="p">,</span> <span class="n">subsMat</span> <span class="o">=</span> <span class="n">rndSubsMat</span><span class="p">,</span> <span class="n">distanceMatrix</span> <span class="o">=</span> <span class="n">distanceMatrix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">rndinfile</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&quot;Checking randomization input file...&quot;</span>
      <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">readFromFile</span><span class="p">(</span><span class="n">rndinfile</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">rndOptions</span><span class="o">.</span><span class="n">number</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&#39;Starting randomizations...&#39;</span>
      <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">randomize</span><span class="p">(</span><span class="n">rndoutfile</span><span class="p">,</span><span class="n">rndOptions</span><span class="o">.</span><span class="n">number</span><span class="p">,</span><span class="n">rndinfile</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">rndinfiles</span><span class="p">:</span>
      <span class="c"># combine multiple files</span>
      <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">combineFromMultiple</span><span class="p">(</span><span class="n">rndoutfile</span><span class="p">,</span><span class="n">rndinfiles</span><span class="p">)</span>


  <span class="c"># Write report</span>
  <span class="k">if</span> <span class="n">outfile</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;&quot;</span>
    <span class="k">print</span> <span class="s">&quot;Preparing report:&quot;</span>

    <span class="k">if</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="o">==</span><span class="n">scoreOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="p">:</span>
      <span class="n">reportSubsetRecords</span> <span class="o">=</span> <span class="n">scoreSubsetRecords</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&quot;Subsetting sequences by report regex </span><span class="si">%s</span><span class="s">...&quot;</span><span class="o">%</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">subsetRE</span>
      <span class="n">reportSubsetRecords</span> <span class="o">=</span> <span class="n">subsetByID</span><span class="p">(</span><span class="n">ALLrecords</span><span class="p">,</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">subsetRE</span><span class="p">)</span>
      <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&quot;List of sequences in output subset:&quot;</span>
        <span class="n">printRecordNames</span><span class="p">(</span><span class="n">reportSubsetRecords</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">alignmentPDF</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&quot;Preparing randomization file </span><span class="si">%s</span><span class="s"> for lookup:&quot;</span><span class="o">%</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">rndlookupfile</span>
      <span class="n">lookupFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">reportOptions</span><span class="o">.</span><span class="n">rndlookupfile</span><span class="p">,</span> <span class="s">&quot;rb&quot;</span><span class="p">)</span>
      <span class="n">alignmentPDF</span><span class="o">.</span><span class="n">prepareLookup</span><span class="p">(</span><span class="n">lookupFile</span><span class="p">)</span>

    <span class="n">generateGenericReport</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="n">theScores</span><span class="p">,</span> <span class="n">records</span> <span class="o">=</span> <span class="n">reportSubsetRecords</span><span class="p">,</span> <span class="n">alignmentPDF</span> <span class="o">=</span> <span class="n">alignmentPDF</span><span class="p">,</span> <span class="n">referenceId</span> <span class="o">=</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">refSeqID</span><span class="p">,</span>
      <span class="n">outputPositions</span> <span class="o">=</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">outputPositions</span><span class="p">,</span> <span class="n">simpleNumbering</span> <span class="o">=</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">simpleNumbering</span><span class="p">,</span> <span class="n">numberingStart</span> <span class="o">=</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">numStart</span><span class="p">,</span>
      <span class="n">blockSequenceOutput</span> <span class="o">=</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">blockOutput</span><span class="p">,</span> <span class="n">showCompleteProfile</span> <span class="o">=</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">showCompleteProfile</span><span class="p">,</span> <span class="n">pAsFraction</span> <span class="o">=</span> <span class="n">reportOptions</span><span class="o">.</span><span class="n">pAsFraction</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;No report requested. Did you forget the -o option?&quot;</span>

  <span class="k">print</span> <span class="s">&quot;Complete!&quot;</span>
</div>
<span class="n">verbose</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">version</span> <span class="o">=</span> <span class="s">&quot;v1.0.8&quot;</span>

<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">profscore: Profile Conservation Scoring 1.0.8 documentation</a> &raquo;</li>
          <li><a href="index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011-2014, Tet Woo Lee.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>